{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f3b6a87",
   "metadata": {},
   "source": [
    "# Introduction to CPTAC and Proteomics\n",
    "\n",
    "This document will take you through the basics of CPTAC and analyzing proteomic data in python. Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ad08b5",
   "metadata": {},
   "source": [
    "## Setting up your working directory\n",
    "\n",
    "Set your working directory to your analysis_data folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a53acaad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.chdir('/Users/amirarya/Desktop/qbio_490_amir/analysis_data')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf734ad",
   "metadata": {},
   "source": [
    "## (1) Introduction to CPTAC\n",
    "\n",
    "To install CPTAC, we will use the `cptac` package. CPTAC, which stands for Clinical Proteomics Tumor Analysis Consortium, is a bit like TCGA in that it is publicly available big data for many different cancer but it focuses mostly on proteomics and proteogenomic data. One cool thing about CPTAC is that its proteomics data is *in conjunction* with TCGA clinical data, so CPTAC and TCGA share patients and we can extract more about their protein expression as an added layer of analysis using CPTAC.\n",
    "\n",
    "The thing about proteomic data is that it doesn't use absolute counts like RNA data. Instead, proteomics data shows the **relative expression/intensity** instead of counts. This makes it easier to compare protein expression data among different groups/demographics of interest because we have a reference value. \n",
    "\n",
    "### Installing and importing `cptac` &#8592; we just did this!\n",
    "The `cptac` package is similar to `TCGAbiolinks`. The information in `cptac` is stored using pandas dataframes. Just like in R, we need to install `cptac` before loading it into our environment. If you haven't installed it yet, just run the code chunk below. If you get an error, you will have to take the following steps:\n",
    "\n",
    "1. Shut down jupyter notebook in the terminal with CTRL+C.\n",
    "2. Make sure that you have deactivated your base environment and activated your qbio_student environment.\n",
    "3. Run `pip install cptac`.\n",
    "4. Open the tutorial again through jupyter.\n",
    "\n",
    "### Start exploring CPTAC with `cptac`\n",
    "Similar to TCGAbiolinks, we need to load the package and download the data before using.\n",
    "1. Import the package (`cptac`).\n",
    "2. Look at the data sets available to us with the list_datasets() function. As a reminder, remember the package_name.method() syntax!\n",
    "3. Download the one of interest (BRCA in our case).\n",
    "4. Load it into our python environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3274cad7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Data reuse status</th>\n",
       "      <th>Publication link</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Brca</th>\n",
       "      <td>breast cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33212010/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ccrcc</th>\n",
       "      <td>clear cell renal cell carcinoma (kidney)</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/31675502/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Colon</th>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/31031003/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endometrial</th>\n",
       "      <td>endometrial carcinoma (uterine)</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/32059776/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gbm</th>\n",
       "      <td>glioblastoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33577785/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hnscc</th>\n",
       "      <td>head and neck squamous cell carcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33417831/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lscc</th>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/34358469/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luad</th>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/32649874/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ovarian</th>\n",
       "      <td>high grade serous ovarian cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/27372738/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pdac</th>\n",
       "      <td>pancreatic ductal adenocarcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/34534465/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UcecConf</th>\n",
       "      <td>endometrial confirmatory carcinoma</td>\n",
       "      <td>password access only</td>\n",
       "      <td>unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GbmConf</th>\n",
       "      <td>glioblastoma confirmatory</td>\n",
       "      <td>password access only</td>\n",
       "      <td>unpublished</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Description     Data reuse status  \\\n",
       "Dataset name                                                                   \n",
       "Brca                                     breast cancer       no restrictions   \n",
       "Ccrcc         clear cell renal cell carcinoma (kidney)       no restrictions   \n",
       "Colon                                colorectal cancer       no restrictions   \n",
       "Endometrial            endometrial carcinoma (uterine)       no restrictions   \n",
       "Gbm                                       glioblastoma       no restrictions   \n",
       "Hnscc            head and neck squamous cell carcinoma       no restrictions   \n",
       "Lscc                      lung squamous cell carcinoma       no restrictions   \n",
       "Luad                               lung adenocarcinoma       no restrictions   \n",
       "Ovarian               high grade serous ovarian cancer       no restrictions   \n",
       "Pdac                  pancreatic ductal adenocarcinoma       no restrictions   \n",
       "UcecConf            endometrial confirmatory carcinoma  password access only   \n",
       "GbmConf                      glioblastoma confirmatory  password access only   \n",
       "\n",
       "                                       Publication link  \n",
       "Dataset name                                             \n",
       "Brca          https://pubmed.ncbi.nlm.nih.gov/33212010/  \n",
       "Ccrcc         https://pubmed.ncbi.nlm.nih.gov/31675502/  \n",
       "Colon         https://pubmed.ncbi.nlm.nih.gov/31031003/  \n",
       "Endometrial   https://pubmed.ncbi.nlm.nih.gov/32059776/  \n",
       "Gbm           https://pubmed.ncbi.nlm.nih.gov/33577785/  \n",
       "Hnscc         https://pubmed.ncbi.nlm.nih.gov/33417831/  \n",
       "Lscc          https://pubmed.ncbi.nlm.nih.gov/34358469/  \n",
       "Luad          https://pubmed.ncbi.nlm.nih.gov/32649874/  \n",
       "Ovarian       https://pubmed.ncbi.nlm.nih.gov/27372738/  \n",
       "Pdac          https://pubmed.ncbi.nlm.nih.gov/34534465/  \n",
       "UcecConf                                    unpublished  \n",
       "GbmConf                                     unpublished  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Import cptac\n",
    "import cptac as cp\n",
    "# 2. Examine the data sets available with list_datasets()\n",
    "cp.list_datasets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "aeb53aa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         \r"
     ]
    }
   ],
   "source": [
    "# 3. Download the breast cancer data set -- fill in dataset!\n",
    "cptac.download(dataset=\"BRCA\")\n",
    "\n",
    "# 4. Load the breast cancer data set\n",
    "brca = cptac.Brca()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fadf54d4",
   "metadata": {},
   "source": [
    "Now, we have our brca object containing a bunch of data. We can use the list_data() function the data available for the breast cancer dataset.\n",
    "\n",
    "In addition to the proteomics data, we can also accesss the accompanying transcriptomics, CNV, etc. using this object. We will focus on the broad proteomics data for this course; however, other data present in this dataset are also pretty interesting. Finally, many proteins can't be identified very well, so they won't appear as a column (for example, APC isn't in the proteomics data set).\n",
    "\n",
    "**Exercise 1.1** Call the list_data() function from the brca object to examine the different data contained in the brca object. Much like calling functions from a package, we can use the brca.function() syntax.\n",
    "\n",
    "* How many patients do you think there are in this dataset? (Make an educated guess about this from the dimensions - we'll confirm your answer later on)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f96afeaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Below are the dataframes contained in this dataset and their dimensions:\n",
      "\n",
      "acetylproteomics\n",
      "\t122 rows\n",
      "\t9868 columns\n",
      "clinical\n",
      "\t122 rows\n",
      "\t18 columns\n",
      "CNV\n",
      "\t122 rows\n",
      "\t23692 columns\n",
      "derived_molecular\n",
      "\t122 rows\n",
      "\t36 columns\n",
      "phosphoproteomics\n",
      "\t122 rows\n",
      "\t38775 columns\n",
      "proteomics\n",
      "\t122 rows\n",
      "\t10107 columns\n",
      "somatic_mutation\n",
      "\t24106 rows\n",
      "\t3 columns\n",
      "transcriptomics\n",
      "\t122 rows\n",
      "\t23121 columns\n"
     ]
    }
   ],
   "source": [
    "# Run the list_data() function\n",
    "brca.list_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab6034bd",
   "metadata": {},
   "source": [
    "**Exercise 1.2** Use the get_proteomics() function to store the proteomics data into a new data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2f7abc05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAED1</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMDC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN31</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>-0.6712</td>\n",
       "      <td>-0.2075</td>\n",
       "      <td>2.7959</td>\n",
       "      <td>1.3969</td>\n",
       "      <td>-1.0899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.6708</td>\n",
       "      <td>-0.3484</td>\n",
       "      <td>-0.4756</td>\n",
       "      <td>-0.7299</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.2868</td>\n",
       "      <td>-0.6536</td>\n",
       "      <td>0.3384</td>\n",
       "      <td>2.1169</td>\n",
       "      <td>1.3910</td>\n",
       "      <td>-2.1230</td>\n",
       "      <td>0.9136</td>\n",
       "      <td>-0.8082</td>\n",
       "      <td>-1.4793</td>\n",
       "      <td>0.9136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>1.3964</td>\n",
       "      <td>1.3302</td>\n",
       "      <td>-5.0948</td>\n",
       "      <td>0.7674</td>\n",
       "      <td>-1.6845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.1022</td>\n",
       "      <td>-0.5814</td>\n",
       "      <td>0.2916</td>\n",
       "      <td>-2.2857</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.7592</td>\n",
       "      <td>0.4711</td>\n",
       "      <td>0.6018</td>\n",
       "      <td>0.2062</td>\n",
       "      <td>-0.2137</td>\n",
       "      <td>-2.1219</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>2.5814</td>\n",
       "      <td>-0.2852</td>\n",
       "      <td>-0.1074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR001</th>\n",
       "      <td>2.0219</td>\n",
       "      <td>1.6269</td>\n",
       "      <td>-3.2943</td>\n",
       "      <td>0.3352</td>\n",
       "      <td>-1.0739</td>\n",
       "      <td>1.2255</td>\n",
       "      <td>0.2754</td>\n",
       "      <td>-1.1187</td>\n",
       "      <td>-0.0534</td>\n",
       "      <td>-0.2519</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2306</td>\n",
       "      <td>-0.3010</td>\n",
       "      <td>0.3395</td>\n",
       "      <td>-0.5316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4996</td>\n",
       "      <td>0.7622</td>\n",
       "      <td>-1.5607</td>\n",
       "      <td>0.0256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR008</th>\n",
       "      <td>-0.5290</td>\n",
       "      <td>0.3267</td>\n",
       "      <td>1.4342</td>\n",
       "      <td>0.4938</td>\n",
       "      <td>-2.8676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0691</td>\n",
       "      <td>-0.3643</td>\n",
       "      <td>-1.8173</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.1789</td>\n",
       "      <td>0.2695</td>\n",
       "      <td>0.1506</td>\n",
       "      <td>1.0498</td>\n",
       "      <td>0.7546</td>\n",
       "      <td>1.7889</td>\n",
       "      <td>-0.2499</td>\n",
       "      <td>-0.2590</td>\n",
       "      <td>-0.1263</td>\n",
       "      <td>0.3725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR009</th>\n",
       "      <td>1.2556</td>\n",
       "      <td>3.4489</td>\n",
       "      <td>2.8043</td>\n",
       "      <td>-0.2956</td>\n",
       "      <td>-1.7261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.0471</td>\n",
       "      <td>-0.3547</td>\n",
       "      <td>-0.8298</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.3990</td>\n",
       "      <td>-0.2596</td>\n",
       "      <td>0.1898</td>\n",
       "      <td>-0.5010</td>\n",
       "      <td>-0.4189</td>\n",
       "      <td>0.3080</td>\n",
       "      <td>0.5057</td>\n",
       "      <td>0.2181</td>\n",
       "      <td>-0.2288</td>\n",
       "      <td>-0.2750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR001</th>\n",
       "      <td>-0.6610</td>\n",
       "      <td>-0.6402</td>\n",
       "      <td>-4.8578</td>\n",
       "      <td>1.2319</td>\n",
       "      <td>-1.6491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.3074</td>\n",
       "      <td>-0.3074</td>\n",
       "      <td>-0.0266</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.2528</td>\n",
       "      <td>0.5090</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>0.4908</td>\n",
       "      <td>-0.5570</td>\n",
       "      <td>2.3864</td>\n",
       "      <td>0.3764</td>\n",
       "      <td>-0.6974</td>\n",
       "      <td>1.3541</td>\n",
       "      <td>1.1123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR002</th>\n",
       "      <td>-1.3735</td>\n",
       "      <td>0.4227</td>\n",
       "      <td>-4.9553</td>\n",
       "      <td>0.6327</td>\n",
       "      <td>-3.1434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.3071</td>\n",
       "      <td>0.7562</td>\n",
       "      <td>-1.6912</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.3351</td>\n",
       "      <td>0.1548</td>\n",
       "      <td>1.0792</td>\n",
       "      <td>-0.6619</td>\n",
       "      <td>-1.4444</td>\n",
       "      <td>-0.3704</td>\n",
       "      <td>0.4909</td>\n",
       "      <td>0.3938</td>\n",
       "      <td>0.2992</td>\n",
       "      <td>-0.3494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR010</th>\n",
       "      <td>1.1583</td>\n",
       "      <td>0.3329</td>\n",
       "      <td>-5.7358</td>\n",
       "      <td>-0.1658</td>\n",
       "      <td>-2.0413</td>\n",
       "      <td>-1.2433</td>\n",
       "      <td>0.9090</td>\n",
       "      <td>-0.2410</td>\n",
       "      <td>0.6717</td>\n",
       "      <td>-0.1651</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.7054</td>\n",
       "      <td>0.2752</td>\n",
       "      <td>0.8850</td>\n",
       "      <td>-2.6704</td>\n",
       "      <td>-0.9444</td>\n",
       "      <td>-1.9717</td>\n",
       "      <td>0.0650</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>-0.0686</td>\n",
       "      <td>0.1798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X22BR005</th>\n",
       "      <td>0.4948</td>\n",
       "      <td>-1.0986</td>\n",
       "      <td>-8.8314</td>\n",
       "      <td>0.2826</td>\n",
       "      <td>-1.0123</td>\n",
       "      <td>-2.5732</td>\n",
       "      <td>5.7567</td>\n",
       "      <td>1.7644</td>\n",
       "      <td>0.5415</td>\n",
       "      <td>0.1531</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.3936</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>-0.9367</td>\n",
       "      <td>-0.1922</td>\n",
       "      <td>1.2572</td>\n",
       "      <td>1.3220</td>\n",
       "      <td>-1.0698</td>\n",
       "      <td>0.4012</td>\n",
       "      <td>-0.3792</td>\n",
       "      <td>1.2752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X22BR006</th>\n",
       "      <td>0.5049</td>\n",
       "      <td>-0.6582</td>\n",
       "      <td>-7.4699</td>\n",
       "      <td>0.6570</td>\n",
       "      <td>-0.7239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.6420</td>\n",
       "      <td>2.2376</td>\n",
       "      <td>-0.7597</td>\n",
       "      <td>0.1589</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5526</td>\n",
       "      <td>0.3468</td>\n",
       "      <td>1.0655</td>\n",
       "      <td>-1.2607</td>\n",
       "      <td>-2.2866</td>\n",
       "      <td>1.0775</td>\n",
       "      <td>-0.1334</td>\n",
       "      <td>2.5269</td>\n",
       "      <td>1.4294</td>\n",
       "      <td>0.4542</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 10107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name          A1BG     A2M   A2ML1    AAAS    AACS   AADAT   AAED1   AAGAB  \\\n",
       "Patient_ID                                                                   \n",
       "CPT000814  -0.6712 -0.2075  2.7959  1.3969 -1.0899     NaN  1.6708 -0.3484   \n",
       "CPT001846   1.3964  1.3302 -5.0948  0.7674 -1.6845     NaN  2.1022 -0.5814   \n",
       "X01BR001    2.0219  1.6269 -3.2943  0.3352 -1.0739  1.2255  0.2754 -1.1187   \n",
       "X01BR008   -0.5290  0.3267  1.4342  0.4938 -2.8676     NaN     NaN -1.0691   \n",
       "X01BR009    1.2556  3.4489  2.8043 -0.2956 -1.7261     NaN     NaN -2.0471   \n",
       "...            ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "X21BR001   -0.6610 -0.6402 -4.8578  1.2319 -1.6491     NaN     NaN -0.3074   \n",
       "X21BR002   -1.3735  0.4227 -4.9553  0.6327 -3.1434     NaN     NaN  0.3071   \n",
       "X21BR010    1.1583  0.3329 -5.7358 -0.1658 -2.0413 -1.2433  0.9090 -0.2410   \n",
       "X22BR005    0.4948 -1.0986 -8.8314  0.2826 -1.0123 -2.5732  5.7567  1.7644   \n",
       "X22BR006    0.5049 -0.6582 -7.4699  0.6570 -0.7239     NaN  0.6420  2.2376   \n",
       "\n",
       "Name          AAK1   AAMDC  ...  ZSCAN31  ZSWIM8    ZW10  ZWILCH   ZWINT  \\\n",
       "Patient_ID                  ...                                            \n",
       "CPT000814  -0.4756 -0.7299  ...  -5.2868 -0.6536  0.3384  2.1169  1.3910   \n",
       "CPT001846   0.2916 -2.2857  ...  -0.7592  0.4711  0.6018  0.2062 -0.2137   \n",
       "X01BR001   -0.0534 -0.2519  ...      NaN  0.2306 -0.3010  0.3395 -0.5316   \n",
       "X01BR008   -0.3643 -1.8173  ...  -2.1789  0.2695  0.1506  1.0498  0.7546   \n",
       "X01BR009   -0.3547 -0.8298  ...  -2.3990 -0.2596  0.1898 -0.5010 -0.4189   \n",
       "...            ...     ...  ...      ...     ...     ...     ...     ...   \n",
       "X21BR001   -0.3074 -0.0266  ...  -0.2528  0.5090  0.0306  0.4908 -0.5570   \n",
       "X21BR002    0.7562 -1.6912  ...  -3.3351  0.1548  1.0792 -0.6619 -1.4444   \n",
       "X21BR010    0.6717 -0.1651  ...  -0.7054  0.2752  0.8850 -2.6704 -0.9444   \n",
       "X22BR005    0.5415  0.1531  ...  -0.3936 -0.0340 -0.9367 -0.1922  1.2572   \n",
       "X22BR006   -0.7597  0.1589  ...   0.5526  0.3468  1.0655 -1.2607 -2.2866   \n",
       "\n",
       "Name          ZXDC  ZYG11B     ZYX   ZZEF1    ZZZ3  \n",
       "Patient_ID                                          \n",
       "CPT000814  -2.1230  0.9136 -0.8082 -1.4793  0.9136  \n",
       "CPT001846  -2.1219  0.0860  2.5814 -0.2852 -0.1074  \n",
       "X01BR001       NaN  0.4996  0.7622 -1.5607  0.0256  \n",
       "X01BR008    1.7889 -0.2499 -0.2590 -0.1263  0.3725  \n",
       "X01BR009    0.3080  0.5057  0.2181 -0.2288 -0.2750  \n",
       "...            ...     ...     ...     ...     ...  \n",
       "X21BR001    2.3864  0.3764 -0.6974  1.3541  1.1123  \n",
       "X21BR002   -0.3704  0.4909  0.3938  0.2992 -0.3494  \n",
       "X21BR010   -1.9717  0.0650  0.6300 -0.0686  0.1798  \n",
       "X22BR005    1.3220 -1.0698  0.4012 -0.3792  1.2752  \n",
       "X22BR006    1.0775 -0.1334  2.5269  1.4294  0.4542  \n",
       "\n",
       "[122 rows x 10107 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data = brca.get_proteomics() # get the proteomics data\n",
    " # view the data\n",
    "\n",
    "\n",
    "# Uncomment and run this command after you get the data, you don't really need to know what it does but it collapses\n",
    "# the column names and gets rid of the database_IDs which will make our lives much easier!\n",
    "\n",
    "protein_data.columns = protein_data.columns.get_level_values(0) \n",
    "protein_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1188f104",
   "metadata": {},
   "source": [
    "## (2) Exploring the Data\n",
    "\n",
    "As mentioned earlier, `cptac` data is in the form of `pandas` dataframes. Let's load that and `numpy`. \n",
    "\n",
    "**Exercise 2.1** Import numpy and pandas with their usual shortened names (pd and np)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ace4453",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b18a548",
   "metadata": {},
   "source": [
    "Remember that we can access the rows/column names of a data frame in two ways:\n",
    "1. We can get the row and column names together in a list with the `axes` attribute.\n",
    "2. We can get the row or column names individually in the `index` (row) or `columns` (column) attribute.\n",
    "\n",
    "**Exercise 2.2** Print the axes, index, and columns of the protein_data object. Then, determine its dimensions using the shape attribute. Finally, answer the following questions:\n",
    "\n",
    "* How many rows and columns are there in the data frame? There are 122 rows and 10107 colums\n",
    "* What do the rows and columns represent? the rows represent the patients and the collums represnets the name of the gene\n",
    "* How many patients are there? 122\n",
    "* What about proteins? 10107"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1501f26c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Index(['CPT000814', 'CPT001846', 'X01BR001', 'X01BR008', 'X01BR009',\n",
      "       'X01BR010', 'X01BR015', 'X01BR017', 'X01BR018', 'X01BR020',\n",
      "       ...\n",
      "       'X20BR002', 'X20BR005', 'X20BR006', 'X20BR007', 'X20BR008', 'X21BR001',\n",
      "       'X21BR002', 'X21BR010', 'X22BR005', 'X22BR006'],\n",
      "      dtype='object', name='Patient_ID', length=122), Index(['A1BG', 'A2M', 'A2ML1', 'AAAS', 'AACS', 'AADAT', 'AAED1', 'AAGAB',\n",
      "       'AAK1', 'AAMDC',\n",
      "       ...\n",
      "       'ZSCAN31', 'ZSWIM8', 'ZW10', 'ZWILCH', 'ZWINT', 'ZXDC', 'ZYG11B', 'ZYX',\n",
      "       'ZZEF1', 'ZZZ3'],\n",
      "      dtype='object', name='Name', length=10107)]\n",
      "Index(['CPT000814', 'CPT001846', 'X01BR001', 'X01BR008', 'X01BR009',\n",
      "       'X01BR010', 'X01BR015', 'X01BR017', 'X01BR018', 'X01BR020',\n",
      "       ...\n",
      "       'X20BR002', 'X20BR005', 'X20BR006', 'X20BR007', 'X20BR008', 'X21BR001',\n",
      "       'X21BR002', 'X21BR010', 'X22BR005', 'X22BR006'],\n",
      "      dtype='object', name='Patient_ID', length=122)\n",
      "Index(['A1BG', 'A2M', 'A2ML1', 'AAAS', 'AACS', 'AADAT', 'AAED1', 'AAGAB',\n",
      "       'AAK1', 'AAMDC',\n",
      "       ...\n",
      "       'ZSCAN31', 'ZSWIM8', 'ZW10', 'ZWILCH', 'ZWINT', 'ZXDC', 'ZYG11B', 'ZYX',\n",
      "       'ZZEF1', 'ZZZ3'],\n",
      "      dtype='object', name='Name', length=10107)\n"
     ]
    }
   ],
   "source": [
    "print(protein_data.axes)\n",
    "print(protein_data.index)\n",
    "print(protein_data.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6db06758",
   "metadata": {},
   "source": [
    "Remember we can access data by name using `.loc[]` and `.iloc[]`. Since this dataset is really large, we won't really use `.iloc[]` which accesses by index. To get the protein expression information for a specific patient, therefore, we would run something like this: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5158710b",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'name'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexes/base.py:3803\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3803\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'name'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mprotein_data\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mZYX\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexing.py:1067\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1065\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1066\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1067\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_tuple\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1068\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1069\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1070\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexing.py:1247\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1245\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1246\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_expand_ellipsis(tup)\n\u001b[0;32m-> 1247\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_lowerdim\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtup\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1249\u001b[0m \u001b[38;5;66;03m# no multi-index, so validate all of the indexers\u001b[39;00m\n\u001b[1;32m   1250\u001b[0m tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_tuple_indexer(tup)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexing.py:967\u001b[0m, in \u001b[0;36m_LocationIndexer._getitem_lowerdim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m    963\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, key \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(tup):\n\u001b[1;32m    964\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_label_like(key):\n\u001b[1;32m    965\u001b[0m         \u001b[38;5;66;03m# We don't need to check for tuples here because those are\u001b[39;00m\n\u001b[1;32m    966\u001b[0m         \u001b[38;5;66;03m#  caught by the _is_nested_tuple_indexer check above.\u001b[39;00m\n\u001b[0;32m--> 967\u001b[0m         section \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mi\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    969\u001b[0m         \u001b[38;5;66;03m# We should never have a scalar section here, because\u001b[39;00m\n\u001b[1;32m    970\u001b[0m         \u001b[38;5;66;03m#  _getitem_lowerdim is only called after a check for\u001b[39;00m\n\u001b[1;32m    971\u001b[0m         \u001b[38;5;66;03m#  is_scalar_access, which that would be.\u001b[39;00m\n\u001b[1;32m    972\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m section\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mndim:\n\u001b[1;32m    973\u001b[0m             \u001b[38;5;66;03m# we're in the middle of slicing through a MultiIndex\u001b[39;00m\n\u001b[1;32m    974\u001b[0m             \u001b[38;5;66;03m# revise the key wrt to `section` by inserting an _NS\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexing.py:1312\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1310\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1311\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1312\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexing.py:1260\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1258\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: \u001b[38;5;28mint\u001b[39m):\n\u001b[1;32m   1259\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1260\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/generic.py:4056\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4054\u001b[0m             new_index \u001b[38;5;241m=\u001b[39m index[loc]\n\u001b[1;32m   4055\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4056\u001b[0m     loc \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4058\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(loc, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   4059\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m loc\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mbool_:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/qbio490_amir/lib/python3.9/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3803\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3810\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'name'"
     ]
    }
   ],
   "source": [
    "protein_data.loc[\"patient ID\",\"protein name\"] # this is the general case, you have to fill in actual IDs/protein names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4e080d",
   "metadata": {},
   "source": [
    "**Exercise 2.3** Remember that the numerical data in CPTAC is the relative fold change, not an absolute value/count. Given this, what kind of number would represent lower relative expression? What kind of number would represent higher relative expression? (Hint: reference level is 0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98e77185",
   "metadata": {},
   "outputs": [],
   "source": [
    "## under 0 would represent a lower relative expression and above 0 would represent a higher relative expresion Answer question in comment form here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccdaaa65",
   "metadata": {},
   "source": [
    "**Exericse 2.4**\n",
    "\n",
    "Print the relative fold change of patient `X01BR008` in the `TP53` protein. Then, interpret this value -- does this patient have higher or lower protein levels than the reference? (Hint: use .loc[] and remember that a `:` can be used in either the row or column slot to access every row or every column.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "112a5055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.9239"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data.loc[\"X01BR008\", \"TP53\"]## write code here This protein has higher levels of protein than refrence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66937a6",
   "metadata": {},
   "source": [
    "### Boolean indexing\n",
    "\n",
    "In R, we used the `ifelse()` function very heavily to assign categories, such as classifying patients as \"old\" or \"young\". In Python, we have the `where(COND, TRUE_VAL, FALSE_VAL)` function from numpy, which works extremely similarly to ifelse(). That is:\n",
    "\n",
    "   1. The first parameter is the condition (a boolean vector; i.e. Trues and Falses).\n",
    "   2. The second parameter is the value to fill the True conditions.\n",
    "   3. The third parameter is the value to fil lthe False conditions.\n",
    "    \n",
    "The following example (split into the four code blocks below) shows how to apply the `where()` function, in the context of telling if a number is odd or even."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "877c4f62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nums</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   nums\n",
       "0     0\n",
       "1     1\n",
       "2     2\n",
       "3     3\n",
       "4     4\n",
       "5     5"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first let's create our example array\n",
    "\n",
    "number_example_array = pd.DataFrame({\"nums\": [0, 1, 2, 3, 4, 5]})\n",
    "number_example_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "877ea8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nums</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    nums\n",
       "0  False\n",
       "1   True\n",
       "2  False\n",
       "3   True\n",
       "4  False\n",
       "5   True"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now, let's make a boolean vector where odds will be True, evens will be False\n",
    "boolean_vector = (number_example_array % 2 == 1)\n",
    "boolean_vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11f0d23e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# then add this vector as a column in our array (we could do this and the above in one step)\n",
    "number_example_array[\"boolean_vector\"] = boolean_vector\n",
    "number_example_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e85d10e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nums</th>\n",
       "      <th>parity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Even</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Odd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Even</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Odd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Even</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>Odd</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   nums parity\n",
       "0     0   Even\n",
       "1     1    Odd\n",
       "2     2   Even\n",
       "3     3    Odd\n",
       "4     4   Even\n",
       "5     5    Odd"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly, let's add a third column called \"parity\" that says based off of \"boolean_vector\" if our value is odd/even\n",
    "number_example_array[\"parity\"] = np.where(boolean_vector, \"Odd\", \"Even\")\n",
    "number_example_array"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae3e913",
   "metadata": {},
   "source": [
    "**Exercise 2.5** Access the following:\n",
    "\n",
    "1. Use boolean indexing to access the patient IDs with high (â‰¥1) `TP53` expression. Save this in a pandas DataFrame called `high_TP53`. HINTS:\n",
    "    * Think about if protein expression information will be contained in a row or a column!\n",
    "    * Create a boolean mask with the row/column that contains the expression information of `TP53`.\n",
    "    * Apply the mask to the row/column names that contains the patient IDs.\n",
    "2. How many patients are there in `high_TP53`? Which axis (columns or rows) represents proteins?\n",
    "3. Add a column to `high_TP53` that has the values \"High\"/\"Average\"/\"Low\" based on ARF1 expression.\n",
    "4. How many patients have high expression of both `TP53` and average expression of `ARF1`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d71d7b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n"
     ]
    }
   ],
   "source": [
    "expression_mask = protein_data.loc[:, 'TP53'] >= 1\n",
    "high_TP53 = pd.DataFrame(protein_data.loc[expression_mask, :])\n",
    "print(len(high_TP53.index))\n",
    "high_TP53[\"ARF1_expression\"] = np.where(high_TP53.loc[:, \"ARF1\"] >= 1, \"High\", np.where(high_TP53.loc[:, \"ARF1\"] <= -1, \"Low\", \"Average\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e91a353",
   "metadata": {},
   "source": [
    "## (3) Additional data in CPTAC\n",
    "\n",
    "You can also access other data using the `cptac` python package for the same patients, using the corresponding get function.\n",
    "\n",
    "**Exercise 3.1** Access the RNA (transcriptomics) and clinical data from the brca object. The function name is very similar in syntax to the get_proteomics() function from before!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "26ae7024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Replicate_Measurement_IDs</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>TMT.Plex</th>\n",
       "      <th>TMT.Channel</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Ischemia.Time.in.Minutes</th>\n",
       "      <th>PAM50</th>\n",
       "      <th>NMF.Cluster</th>\n",
       "      <th>NMF.Cluster.Membership.Score</th>\n",
       "      <th>Age.in.Month</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>ER.Updated.Clinical.Status</th>\n",
       "      <th>PR.Clinical.Status</th>\n",
       "      <th>ERBB2.Updated.Clinical.Status</th>\n",
       "      <th>TNBC.Updated.Clinical.Status</th>\n",
       "      <th>ERBB2.Proteogenomic.Status</th>\n",
       "      <th>TOP2A.Proteogenomic.Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>CPT000814</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>13</td>\n",
       "      <td>127C</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>black.or.african.american</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>CPT001846</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>12</td>\n",
       "      <td>128C</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR001</th>\n",
       "      <td>X01BR001</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>2</td>\n",
       "      <td>129N</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.782</td>\n",
       "      <td>660.0</td>\n",
       "      <td>female</td>\n",
       "      <td>black.or.african.american</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR008</th>\n",
       "      <td>X01BR008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>127C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR009</th>\n",
       "      <td>X01BR009</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>127N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR001</th>\n",
       "      <td>X21BR001</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>128N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LumB</td>\n",
       "      <td>LumB-I</td>\n",
       "      <td>0.536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR002</th>\n",
       "      <td>X21BR002</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>128C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LumA</td>\n",
       "      <td>LumB-I</td>\n",
       "      <td>0.443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>positive</td>\n",
       "      <td>positive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X21BR010</th>\n",
       "      <td>X21BR010|X21BR010.REP1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>3|17</td>\n",
       "      <td>129C|128C</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>18.0</td>\n",
       "      <td>LumA</td>\n",
       "      <td>LumA-I</td>\n",
       "      <td>0.559</td>\n",
       "      <td>852.0</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>positive</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X22BR005</th>\n",
       "      <td>X22BR005</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>6</td>\n",
       "      <td>129N</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>20.0</td>\n",
       "      <td>LumA</td>\n",
       "      <td>LumB-I</td>\n",
       "      <td>0.392</td>\n",
       "      <td>552.0</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>positive</td>\n",
       "      <td>positive</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X22BR006</th>\n",
       "      <td>X22BR006</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>10</td>\n",
       "      <td>130N</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>25.0</td>\n",
       "      <td>LumA</td>\n",
       "      <td>LumB-I</td>\n",
       "      <td>0.402</td>\n",
       "      <td>660.0</td>\n",
       "      <td>female</td>\n",
       "      <td>black.or.african.american</td>\n",
       "      <td>positive</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Replicate_Measurement_IDs Sample_Tumor_Normal TMT.Plex TMT.Channel  \\\n",
       "Patient_ID                                                                      \n",
       "CPT000814                  CPT000814               Tumor       13        127C   \n",
       "CPT001846                  CPT001846               Tumor       12        128C   \n",
       "X01BR001                    X01BR001               Tumor        2        129N   \n",
       "X01BR008                    X01BR008               Tumor       16        127C   \n",
       "X01BR009                    X01BR009               Tumor       16        127N   \n",
       "...                              ...                 ...      ...         ...   \n",
       "X21BR001                    X21BR001               Tumor       16        128N   \n",
       "X21BR002                    X21BR002               Tumor       16        128C   \n",
       "X21BR010      X21BR010|X21BR010.REP1               Tumor     3|17   129C|128C   \n",
       "X22BR005                    X22BR005               Tumor        6        129N   \n",
       "X22BR006                    X22BR006               Tumor       10        130N   \n",
       "\n",
       "Name             Stage  Ischemia.Time.in.Minutes  PAM50 NMF.Cluster  \\\n",
       "Patient_ID                                                            \n",
       "CPT000814    Stage IIA                       NaN  Basal     Basal-I   \n",
       "CPT001846    Stage III                       NaN  Basal     Basal-I   \n",
       "X01BR001     Stage IIB                       0.0  Basal     Basal-I   \n",
       "X01BR008           NaN                       NaN  Basal     Basal-I   \n",
       "X01BR009           NaN                       NaN  Basal     Basal-I   \n",
       "...                ...                       ...    ...         ...   \n",
       "X21BR001           NaN                       NaN   LumB      LumB-I   \n",
       "X21BR002           NaN                       NaN   LumA      LumB-I   \n",
       "X21BR010     Stage IIA                      18.0   LumA      LumA-I   \n",
       "X22BR005     Stage IIB                      20.0   LumA      LumB-I   \n",
       "X22BR006    Stage IIIA                      25.0   LumA      LumB-I   \n",
       "\n",
       "Name        NMF.Cluster.Membership.Score  Age.in.Month  Gender  \\\n",
       "Patient_ID                                                       \n",
       "CPT000814                          1.000           NaN     NaN   \n",
       "CPT001846                          0.672           NaN     NaN   \n",
       "X01BR001                           0.782         660.0  female   \n",
       "X01BR008                           0.958           NaN     NaN   \n",
       "X01BR009                           0.825           NaN     NaN   \n",
       "...                                  ...           ...     ...   \n",
       "X21BR001                           0.536           NaN     NaN   \n",
       "X21BR002                           0.443           NaN     NaN   \n",
       "X21BR010                           0.559         852.0  female   \n",
       "X22BR005                           0.392         552.0  female   \n",
       "X22BR006                           0.402         660.0  female   \n",
       "\n",
       "Name                        Ethnicity ER.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                         \n",
       "CPT000814   black.or.african.american                   negative   \n",
       "CPT001846                       white                   negative   \n",
       "X01BR001    black.or.african.american                   negative   \n",
       "X01BR008                          NaN                        NaN   \n",
       "X01BR009                          NaN                   negative   \n",
       "...                               ...                        ...   \n",
       "X21BR001                        white                   negative   \n",
       "X21BR002                        white                   positive   \n",
       "X21BR010                        white                   positive   \n",
       "X22BR005                        white                   positive   \n",
       "X22BR006    black.or.african.american                   positive   \n",
       "\n",
       "Name       PR.Clinical.Status ERBB2.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                    \n",
       "CPT000814            negative                           NaN   \n",
       "CPT001846            negative                           NaN   \n",
       "X01BR001             negative                      negative   \n",
       "X01BR008                  NaN                           NaN   \n",
       "X01BR009             negative                           NaN   \n",
       "...                       ...                           ...   \n",
       "X21BR001             negative                           NaN   \n",
       "X21BR002             positive                           NaN   \n",
       "X21BR010             positive                      negative   \n",
       "X22BR005             positive                      positive   \n",
       "X22BR006             positive                      negative   \n",
       "\n",
       "Name       TNBC.Updated.Clinical.Status ERBB2.Proteogenomic.Status  \\\n",
       "Patient_ID                                                           \n",
       "CPT000814                      positive                   negative   \n",
       "CPT001846                      positive                   negative   \n",
       "X01BR001                       positive                   negative   \n",
       "X01BR008                            NaN                   negative   \n",
       "X01BR009                       positive                   negative   \n",
       "...                                 ...                        ...   \n",
       "X21BR001                       negative                   positive   \n",
       "X21BR002                       negative                   negative   \n",
       "X21BR010                       negative                   negative   \n",
       "X22BR005                       negative                   positive   \n",
       "X22BR006                       negative                   negative   \n",
       "\n",
       "Name       TOP2A.Proteogenomic.Status  \n",
       "Patient_ID                             \n",
       "CPT000814                    negative  \n",
       "CPT001846                    negative  \n",
       "X01BR001                     negative  \n",
       "X01BR008                     negative  \n",
       "X01BR009                     negative  \n",
       "...                               ...  \n",
       "X21BR001                     negative  \n",
       "X21BR002                     negative  \n",
       "X21BR010                     negative  \n",
       "X22BR005                     negative  \n",
       "X22BR006                     negative  \n",
       "\n",
       "[122 rows x 18 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_data = brca.get_transcriptomics()\n",
    "clinical_data = brca.get_clinical()\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d217ca6",
   "metadata": {},
   "source": [
    "**Exercise 3.2** Examine these dataframes. In particular:\n",
    "\n",
    "* What do the values in the rna_data data frame represent?\n",
    "* Compare the dimensions of the high_TP53, rna_data, and clinical_data. Are they equal?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ef35657",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "67f6107d",
   "metadata": {},
   "source": [
    "The dimensions for each dataframe might not be equal depending on if you masked some patients. That's where the `intersect1d()` function from numpy becomes really useful, especially if we want to analyze transcriptomics, proteomics, and clinical data together. This lets us see which patient barcodes are shared between the data frames.\n",
    "\n",
    "\n",
    "Let's take a look at all of the patients who are older than 50 years old at the time of diagnosis. \n",
    "**Exercise 3.3**\n",
    "1.  Use the clinical data DataFrame to create a boolean mask for the condition of patients being older than 50.\n",
    "2.  Mask the clinical data and assign it to a DataFrame called 'masked_clinical'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "27343c31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Patient_ID\n",
       "CPT000814   -0.6712\n",
       "CPT001846    1.3964\n",
       "X01BR001     2.0219\n",
       "X01BR008    -0.5290\n",
       "X01BR009     1.2556\n",
       "              ...  \n",
       "X21BR001    -0.6610\n",
       "X21BR002    -1.3735\n",
       "X21BR010     1.1583\n",
       "X22BR005     0.4948\n",
       "X22BR006     0.5049\n",
       "Name: A1BG, Length: 122, dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_mask = clinical_data.loc[:, 'Age.in.Month'] > 50 * 12\n",
    "\n",
    "masked_clinical = pd.DataFrame(clinical_data.loc[old_mask, :])\n",
    "\n",
    "protein_data.iloc[:, 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d356f73c",
   "metadata": {},
   "source": [
    "Now that we have masked our data, if we tried to do an analysis using the different data frames, we would get a lot of errors since we are now missing some patients in our masked_clinical data frame. If we want to get the patients who are present in the masked data as well as the rna_data and protein_data, we can use intersect1d().\n",
    "\n",
    "**Exercise 3.4**\n",
    "1. Use `intersect1d()` from np to create a list that contains the patient IDs for each pair of data frames as mentioned below. (Hint: this function takes two 1d arrays, which would be patient IDs from two different dataframes. How do we get this from each dataframe? Are patient IDs rows or columns?) \n",
    "2. Use a for loop to print the lengths of these three data frames, and compare them to the number of rows. The for loop skeleton is below. What do you notice about the three lengths? Do the numbers correspond to anything in particular?\n",
    "3. Finally, identify which intersection contains the patient IDs where there is data for all three levels of data (proteomics, transcriptomics, and clinical). How can we tell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6daf99fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of 0 is 122\n",
      "The length of 1 is 79\n",
      "The length of 2 is 79\n"
     ]
    }
   ],
   "source": [
    "#1. \n",
    "intersects_d = [\n",
    "    np.intersect1d(rna_data.index, protein_data.index),\n",
    "    np.intersect1d(masked_clinical.index, protein_data.index),\n",
    "    np.intersect1d(rna_data.index, masked_clinical.index)\n",
    "]\n",
    "# 2. Print the lengths here\n",
    "# write a for loop here\n",
    "for i in range(len(intersects_d)):\n",
    "    print(f\"The length of {i} is {len(intersects_d[i])}\")\n",
    "\n",
    "    \n",
    "# 3. Which comparison(s) contain the patient names that have all three levels\n",
    "# The length of 1 and 2 is 79."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f34fa008",
   "metadata": {},
   "source": [
    "## (4) Examining the Clinical Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c16ae96",
   "metadata": {},
   "source": [
    "Let's explore the clinical data in more detail.\n",
    "\n",
    "First, use head() to glance as to what data is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "60fe1771",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of Name       Replicate_Measurement_IDs Sample_Tumor_Normal TMT.Plex TMT.Channel  \\\n",
       "Patient_ID                                                                      \n",
       "CPT000814                  CPT000814               Tumor       13        127C   \n",
       "CPT001846                  CPT001846               Tumor       12        128C   \n",
       "X01BR001                    X01BR001               Tumor        2        129N   \n",
       "X01BR008                    X01BR008               Tumor       16        127C   \n",
       "X01BR009                    X01BR009               Tumor       16        127N   \n",
       "...                              ...                 ...      ...         ...   \n",
       "X21BR001                    X21BR001               Tumor       16        128N   \n",
       "X21BR002                    X21BR002               Tumor       16        128C   \n",
       "X21BR010      X21BR010|X21BR010.REP1               Tumor     3|17   129C|128C   \n",
       "X22BR005                    X22BR005               Tumor        6        129N   \n",
       "X22BR006                    X22BR006               Tumor       10        130N   \n",
       "\n",
       "Name             Stage  Ischemia.Time.in.Minutes  PAM50 NMF.Cluster  \\\n",
       "Patient_ID                                                            \n",
       "CPT000814    Stage IIA                       NaN  Basal     Basal-I   \n",
       "CPT001846    Stage III                       NaN  Basal     Basal-I   \n",
       "X01BR001     Stage IIB                       0.0  Basal     Basal-I   \n",
       "X01BR008           NaN                       NaN  Basal     Basal-I   \n",
       "X01BR009           NaN                       NaN  Basal     Basal-I   \n",
       "...                ...                       ...    ...         ...   \n",
       "X21BR001           NaN                       NaN   LumB      LumB-I   \n",
       "X21BR002           NaN                       NaN   LumA      LumB-I   \n",
       "X21BR010     Stage IIA                      18.0   LumA      LumA-I   \n",
       "X22BR005     Stage IIB                      20.0   LumA      LumB-I   \n",
       "X22BR006    Stage IIIA                      25.0   LumA      LumB-I   \n",
       "\n",
       "Name        NMF.Cluster.Membership.Score  Age.in.Month  Gender  \\\n",
       "Patient_ID                                                       \n",
       "CPT000814                          1.000           NaN     NaN   \n",
       "CPT001846                          0.672           NaN     NaN   \n",
       "X01BR001                           0.782         660.0  female   \n",
       "X01BR008                           0.958           NaN     NaN   \n",
       "X01BR009                           0.825           NaN     NaN   \n",
       "...                                  ...           ...     ...   \n",
       "X21BR001                           0.536           NaN     NaN   \n",
       "X21BR002                           0.443           NaN     NaN   \n",
       "X21BR010                           0.559         852.0  female   \n",
       "X22BR005                           0.392         552.0  female   \n",
       "X22BR006                           0.402         660.0  female   \n",
       "\n",
       "Name                        Ethnicity ER.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                         \n",
       "CPT000814   black.or.african.american                   negative   \n",
       "CPT001846                       white                   negative   \n",
       "X01BR001    black.or.african.american                   negative   \n",
       "X01BR008                          NaN                        NaN   \n",
       "X01BR009                          NaN                   negative   \n",
       "...                               ...                        ...   \n",
       "X21BR001                        white                   negative   \n",
       "X21BR002                        white                   positive   \n",
       "X21BR010                        white                   positive   \n",
       "X22BR005                        white                   positive   \n",
       "X22BR006    black.or.african.american                   positive   \n",
       "\n",
       "Name       PR.Clinical.Status ERBB2.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                    \n",
       "CPT000814            negative                           NaN   \n",
       "CPT001846            negative                           NaN   \n",
       "X01BR001             negative                      negative   \n",
       "X01BR008                  NaN                           NaN   \n",
       "X01BR009             negative                           NaN   \n",
       "...                       ...                           ...   \n",
       "X21BR001             negative                           NaN   \n",
       "X21BR002             positive                           NaN   \n",
       "X21BR010             positive                      negative   \n",
       "X22BR005             positive                      positive   \n",
       "X22BR006             positive                      negative   \n",
       "\n",
       "Name       TNBC.Updated.Clinical.Status ERBB2.Proteogenomic.Status  \\\n",
       "Patient_ID                                                           \n",
       "CPT000814                      positive                   negative   \n",
       "CPT001846                      positive                   negative   \n",
       "X01BR001                       positive                   negative   \n",
       "X01BR008                            NaN                   negative   \n",
       "X01BR009                       positive                   negative   \n",
       "...                                 ...                        ...   \n",
       "X21BR001                       negative                   positive   \n",
       "X21BR002                       negative                   negative   \n",
       "X21BR010                       negative                   negative   \n",
       "X22BR005                       negative                   positive   \n",
       "X22BR006                       negative                   negative   \n",
       "\n",
       "Name       TOP2A.Proteogenomic.Status  \n",
       "Patient_ID                             \n",
       "CPT000814                    negative  \n",
       "CPT001846                    negative  \n",
       "X01BR001                     negative  \n",
       "X01BR008                     negative  \n",
       "X01BR009                     negative  \n",
       "...                               ...  \n",
       "X21BR001                     negative  \n",
       "X21BR002                     negative  \n",
       "X21BR010                     negative  \n",
       "X22BR005                     negative  \n",
       "X22BR006                     negative  \n",
       "\n",
       "[122 rows x 18 columns]>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77030e93",
   "metadata": {},
   "source": [
    "**Exercise 4.1** You might notice that the Age column values don't look like years -- they're in months. Create a new column called \"Age\" with the \"Age.in.Month\" values / 12. Then, use head() again to make sure the ages were converted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e2b7a6d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of Name       Replicate_Measurement_IDs Sample_Tumor_Normal TMT.Plex TMT.Channel  \\\n",
       "Patient_ID                                                                      \n",
       "CPT000814                  CPT000814               Tumor       13        127C   \n",
       "CPT001846                  CPT001846               Tumor       12        128C   \n",
       "X01BR001                    X01BR001               Tumor        2        129N   \n",
       "X01BR008                    X01BR008               Tumor       16        127C   \n",
       "X01BR009                    X01BR009               Tumor       16        127N   \n",
       "...                              ...                 ...      ...         ...   \n",
       "X21BR001                    X21BR001               Tumor       16        128N   \n",
       "X21BR002                    X21BR002               Tumor       16        128C   \n",
       "X21BR010      X21BR010|X21BR010.REP1               Tumor     3|17   129C|128C   \n",
       "X22BR005                    X22BR005               Tumor        6        129N   \n",
       "X22BR006                    X22BR006               Tumor       10        130N   \n",
       "\n",
       "Name             Stage  Ischemia.Time.in.Minutes  PAM50 NMF.Cluster  \\\n",
       "Patient_ID                                                            \n",
       "CPT000814    Stage IIA                       NaN  Basal     Basal-I   \n",
       "CPT001846    Stage III                       NaN  Basal     Basal-I   \n",
       "X01BR001     Stage IIB                       0.0  Basal     Basal-I   \n",
       "X01BR008           NaN                       NaN  Basal     Basal-I   \n",
       "X01BR009           NaN                       NaN  Basal     Basal-I   \n",
       "...                ...                       ...    ...         ...   \n",
       "X21BR001           NaN                       NaN   LumB      LumB-I   \n",
       "X21BR002           NaN                       NaN   LumA      LumB-I   \n",
       "X21BR010     Stage IIA                      18.0   LumA      LumA-I   \n",
       "X22BR005     Stage IIB                      20.0   LumA      LumB-I   \n",
       "X22BR006    Stage IIIA                      25.0   LumA      LumB-I   \n",
       "\n",
       "Name        NMF.Cluster.Membership.Score  Age.in.Month  Gender  \\\n",
       "Patient_ID                                                       \n",
       "CPT000814                          1.000           NaN     NaN   \n",
       "CPT001846                          0.672           NaN     NaN   \n",
       "X01BR001                           0.782         660.0  female   \n",
       "X01BR008                           0.958           NaN     NaN   \n",
       "X01BR009                           0.825           NaN     NaN   \n",
       "...                                  ...           ...     ...   \n",
       "X21BR001                           0.536           NaN     NaN   \n",
       "X21BR002                           0.443           NaN     NaN   \n",
       "X21BR010                           0.559         852.0  female   \n",
       "X22BR005                           0.392         552.0  female   \n",
       "X22BR006                           0.402         660.0  female   \n",
       "\n",
       "Name                        Ethnicity ER.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                         \n",
       "CPT000814   black.or.african.american                   negative   \n",
       "CPT001846                       white                   negative   \n",
       "X01BR001    black.or.african.american                   negative   \n",
       "X01BR008                          NaN                        NaN   \n",
       "X01BR009                          NaN                   negative   \n",
       "...                               ...                        ...   \n",
       "X21BR001                        white                   negative   \n",
       "X21BR002                        white                   positive   \n",
       "X21BR010                        white                   positive   \n",
       "X22BR005                        white                   positive   \n",
       "X22BR006    black.or.african.american                   positive   \n",
       "\n",
       "Name       PR.Clinical.Status ERBB2.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                    \n",
       "CPT000814            negative                           NaN   \n",
       "CPT001846            negative                           NaN   \n",
       "X01BR001             negative                      negative   \n",
       "X01BR008                  NaN                           NaN   \n",
       "X01BR009             negative                           NaN   \n",
       "...                       ...                           ...   \n",
       "X21BR001             negative                           NaN   \n",
       "X21BR002             positive                           NaN   \n",
       "X21BR010             positive                      negative   \n",
       "X22BR005             positive                      positive   \n",
       "X22BR006             positive                      negative   \n",
       "\n",
       "Name       TNBC.Updated.Clinical.Status ERBB2.Proteogenomic.Status  \\\n",
       "Patient_ID                                                           \n",
       "CPT000814                      positive                   negative   \n",
       "CPT001846                      positive                   negative   \n",
       "X01BR001                       positive                   negative   \n",
       "X01BR008                            NaN                   negative   \n",
       "X01BR009                       positive                   negative   \n",
       "...                                 ...                        ...   \n",
       "X21BR001                       negative                   positive   \n",
       "X21BR002                       negative                   negative   \n",
       "X21BR010                       negative                   negative   \n",
       "X22BR005                       negative                   positive   \n",
       "X22BR006                       negative                   negative   \n",
       "\n",
       "Name       TOP2A.Proteogenomic.Status   Age  \n",
       "Patient_ID                                   \n",
       "CPT000814                    negative   NaN  \n",
       "CPT001846                    negative   NaN  \n",
       "X01BR001                     negative  55.0  \n",
       "X01BR008                     negative   NaN  \n",
       "X01BR009                     negative   NaN  \n",
       "...                               ...   ...  \n",
       "X21BR001                     negative   NaN  \n",
       "X21BR002                     negative   NaN  \n",
       "X21BR010                     negative  71.0  \n",
       "X22BR005                     negative  46.0  \n",
       "X22BR006                     negative  55.0  \n",
       "\n",
       "[122 rows x 19 columns]>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data['Age'] = clinical_data.loc[:, 'Age.in.Month'] / 12\n",
    "clinical_data.head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7e704a4",
   "metadata": {},
   "source": [
    "There are actually control (non-cancer) tissue samples in the data! Let's show this using the unique() function from numpy.\n",
    "\n",
    "__Exercise 4.2__ Call the unique() function on the Sample_Tumor_Normal column. The function will return the unique values in the column in a sorted order -- this is super useful for examining categorical variables, like tumor stage and tumor status, for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5b56fa09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Tumor'], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "clinical_data.loc[:, 'Sample_Tumor_Normal'].unique()\n",
    "# Get the levels of the Sample_Tumor_Normal column with unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b1338d1",
   "metadata": {},
   "source": [
    "__Exercise 4.3__ Let's examine how tumor stage varies as a function of age:\n",
    "\n",
    "   1. Import the plotting libraries matplotlib.pyplot and seaborn (use the standard abbreviations presented last time).\n",
    "   2. Create an age_category column in clinical_data. As before, we'll define \"Young\" as under 50, and \"Old\" as 50 and older. (Hint: use the where() function from numpy, the equivalent to ifelse()).\n",
    "   3. The \"Stage\" information for these patients is NaN (the pandas version of NA). Use the isna() function from pandas and boolean indexing (i.e. where() from numpy again) to replace these NaN values in the Stage column with \"No cancer\".\n",
    "   4. Using nested for loop and boolean indexing, count the number of old and young patients that have No cancer, Stage I, Stage II, Stage III, and Stage IV cancer.\n",
    "   5. Compare your results to the barplot. Do your numbers make sense?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3d9b0202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Old\n",
      "Category is Old, Stage is No cancer, Number is 15\n",
      "Category is Old, Stage is Stage IA, Number is 4\n",
      "Category is Old, Stage is Stage IIA, Number is 44\n",
      "Category is Old, Stage is Stage IIB, Number is 12\n",
      "Category is Old, Stage is Stage III, Number is 2\n",
      "Category is Old, Stage is Stage IIIA, Number is 18\n",
      "Category is Old, Stage is Stage IIIB, Number is 2\n",
      "Category is Old, Stage is Stage IIIC, Number is 2\n",
      "Young\n",
      "Category is Young, Stage is No cancer, Number is 0\n",
      "Category is Young, Stage is Stage IA, Number is 0\n",
      "Category is Young, Stage is Stage IIA, Number is 6\n",
      "Category is Young, Stage is Stage IIB, Number is 8\n",
      "Category is Young, Stage is Stage III, Number is 2\n",
      "Category is Young, Stage is Stage IIIA, Number is 4\n",
      "Category is Young, Stage is Stage IIIB, Number is 1\n",
      "Category is Young, Stage is Stage IIIC, Number is 2\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjIAAAGwCAYAAACzXI8XAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/av/WaAAAACXBIWXMAAA9hAAAPYQGoP6dpAABRDElEQVR4nO3de1gU590//vfqIqwuyyIiqwZdFEQISrqKBOmXRA4FNQgUzUGtgqI+iUmfhFgNWDWJSYwWxSYxB1EBq9YWD5TUQxQsxqAGNYiJRVRcipZjUcCNchDm94c/5+nGE4eFZcz7dV17XZmZe+77M3CRfTtzz4xMEAQBRERERBLUw9wFEBEREbUXgwwRERFJFoMMERERSRaDDBEREUkWgwwRERFJFoMMERERSRaDDBEREUmW3NwFdLaWlhaUlpbC2toaMpnM3OUQERFRKwiCgBs3bmDgwIHo0ePB510e+yBTWloKR0dHc5dBRERE7XDlyhU88cQTD9z+2AcZa2trAHd+ECqVyszVEBERUWvU1dXB0dFR/B5/kMc+yNy9nKRSqRhkiIiIJOZR00I42ZeIiIgki0GGiIiIJItBhoiIiCTrsZ8jQ0REj4fm5mY0NTWZuwwyEQsLC/Ts2bPD/TDIEBFRtyYIAsrLy1FTU2PuUsjE1Go1NBpNh57zxiBDRETd2t0Q079/f/Tu3ZsPN30MCIKAmzdvorKyEgAwYMCAdvfFIENERN1Wc3OzGGLs7OzMXQ6ZkEKhAABUVlaif//+7b7MxMm+RETUbd2dE9O7d28zV0Kd4e7vtSNznxhkiIio2+PlpMeTKX6vDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQmUFVVhZdffhmDBw+GpaUlNBoNgoODkZOTA+DOfJD09HTzFvkY4u3XREREJhAZGYnGxkakpqZi6NChqKioQFZWFqqrq81d2mONZ2SIiIg6qKamBkePHsWqVaswfvx4DBkyBGPHjkVcXBwmT54MrVYLAIiIiIBMJhOXi4qKEBYWBgcHByiVSnh5eSEzM9Oo77KyMkyaNAkKhQJOTk7Yvn07tFot1q1bZzR+TEwM7O3toVKp4O/vj/z8/C46evPiGRkS+X7s2+lj5LyW0+ljEBF1NaVSCaVSifT0dDz99NOwtLQ02n7y5En0798fycnJCAkJER/+ZjAYMHHiRLz//vuwtLTEli1bEBoaisLCQgwePBgAMHPmTPznP/9BdnY2LCwsEBsbKz4R966pU6dCoVBg//79sLGxwRdffIGAgABcuHABffv27ZofgpnwjAwREVEHyeVypKSkIDU1FWq1Gr6+voiPj8fZs2cBAPb29gD+791Cd5c9PT0xf/58eHh4wMXFBStWrMCwYcOQkZEBADh//jwyMzORlJQEb29v6HQ6bNy4Ebdu3RLH/uabb5Cbm4u0tDSMGTMGLi4uSEhIgFqtxs6dO7v4J9H1GGSIiIhMIDIyEqWlpcjIyEBISAiys7Oh0+mQkpLywH0MBgMWLlwINzc3qNVqKJVKFBQUoKSkBABQWFgIuVwOnU4n7uPs7AxbW1txOT8/HwaDAXZ2duKZIaVSCb1ej6Kiok473u6Cl5aIiIhMxMrKCkFBQQgKCsLSpUsRExOD5cuXIyoq6r7tFy5ciEOHDiEhIQHOzs5QKBSYMmUKGhsbWz2mwWDAgAEDkJ2dfc82tVrdvgOREAYZIiKiTuLu7i7ecm1hYYHm5maj7Tk5OYiKikJERASAO6GkuLhY3O7q6orbt28jLy8Po0ePBgBcunQJ169fF9vodDqUl5dDLpeLk4h/TnhpiYiIqIOqq6vh7++PrVu34uzZs9Dr9UhLS8Pq1asRFhYGANBqtcjKykJ5ebkYRFxcXLB7926cOXMG+fn5mDZtGlpaWsR+R4wYgcDAQMybNw+5ubnIy8vDvHnzoFAoxPcUBQYGwsfHB+Hh4Th48CCKi4tx7NgxLFmyBKdOner6H0YXY5AhIiLqIKVSCW9vbyQmJsLPzw8eHh5YunQp5s6di08++QQAsGbNGhw6dAiOjo74xS9+AQBYu3YtbG1tMW7cOISGhiI4ONhoPgwAbNmyBQ4ODvDz80NERATmzp0La2trWFlZAbjzoL19+/bBz88P0dHRGD58OF588UX861//goODQ9f+IMxAJgiCYO4iOlNdXR1sbGxQW1sLlUpl7nK6Nd5+TUTdTX19PfR6PZycnMQv7p+7q1evwtHREZmZmQgICDB3OR3ysN9va7+/OUeGiIioGzt8+DAMBgNGjhyJsrIyLFq0CFqtFn5+fuYurVtgkCEiIurGmpqaEB8fj8uXL8Pa2hrjxo3Dtm3bYGFhYe7SugUGGSIiom4sODgYwcHB5i6j2+JkXyIiIpIsBhkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLAYZIiIikizefk1ERJIz+ndbunS803+Y2ab2UVFRSE1NxcqVK/HWW2+J69PT0xEREYHH/KH6XYpnZIiIiDqBlZUVVq1aZfSm6p+r5uZmo5dhmhKDDBERUScIDAyERqPBypUrH9pu165dePLJJ2FpaQmtVos1a9Y8su8vv/wSXl5esLKyQr9+/RARESFu+9Of/oQxY8bA2toaGo0G06ZNQ2Vlpbg9OzsbMpkMWVlZGDNmDHr37o1x48ahsLCw1WM0NDRg4cKFGDRoEPr06QNvb29kZ2eL21NSUqBWq5GRkQF3d3dYWlqipKTkkcfVHgwyREREnaBnz5744IMP8PHHH+Pq1av3bXP69Gk8//zzePHFF/H999/j7bffxtKlS5GSkvLAfvfu3YuIiAhMnDgReXl5yMrKwtixY8XtTU1NWLFiBfLz85Geno7i4mJERUXd08+SJUuwZs0anDp1CnK5HLNnz271GK+++iqOHz+OHTt24OzZs5g6dSpCQkJw8eJFsc3NmzexatUqbNy4EefOnUP//v3b8NNrPb79mkR8+zURdTcPejuyFObI1NTUID09HT4+PnB3d8emTZvumSMzffp0VFVV4eDBg+K+ixYtwt69e3Hu3Ln79j1u3DgMHToUW7dubVUtp06dgpeXF27cuAGlUons7GyMHz/e6O3Z+/btw6RJk3Dr1i1YWVk9dIySkhIMHToUJSUlGDhwoLg+MDAQY8eOxQcffICUlBRER0fjzJkz8PT0fGBtpnj7Nc/IEBERdaJVq1YhNTUVBQUF92wrKCiAr6/xPyJ9fX1x8eJFNDc337e/M2fOiAHkfk6fPo3Q0FAMHjwY1tbWeOaZZwDgnks7o0aNEv97wIABACBegnrYGN9//z2am5sxfPhwKJVK8XPkyBEUFRWJ7Xr16mU0RmfhXUtERESdyM/PD8HBwYiLi7vvJZ62UigUD9z2448/ii+Z3LZtG+zt7VFSUoLg4GA0NjYatf3vt2fLZDIAECfkPmwMg8GAnj174vTp0+jZs6fRNqVSaVTn3X47E4MMERFRJ/vwww/x1FNPwdXV1Wi9m5sbcnKML7nn5ORg+PDh94SEu0aNGoWsrCxER0ffs+38+fOorq7Ghx9+CEdHRwB3Li211cPG+MUvfoHm5mZUVlbi//2//9fmvk2NQYaIiKiTjRw5EtOnT8dHH31ktP7NN9+El5cXVqxYgRdeeAHHjx/HJ598gk8//fSBfS1fvhwBAQEYNmwYXnzxRdy+fRv79u3D4sWLMXjwYPTq1Qsff/wx/ud//gc//PADVqxY0eZ6HzbG8OHDMX36dMycORNr1qzBL37xC1RVVSErKwujRo3CpEmT2jxeR3CODBERURd4991373mWik6nw1//+lfs2LEDHh4eWLZsGd59992HXoJ69tlnkZaWhoyMDDz11FPw9/dHbm4uAMDe3h4pKSlIS0uDu7s7PvzwQyQkJLS51oeNAQDJycmYOXMm3nzzTbi6uiI8PBwnT57E4MGD2zxWR/GuJRLxriUi6m4edlcLSd9jddfShx9+CJlMhtdff11cV19fjwULFsDOzg5KpRKRkZGoqKgwX5FERETUrXSLIHPy5El88cUX99ym9cYbb+DLL79EWloajhw5gtLSUvz61782U5VERETU3Zg9yBgMBkyfPh1JSUmwtbUV19fW1mLTpk1Yu3Yt/P39MXr0aCQnJ+PYsWM4ceKEGSsmIiKi7sLsQWbBggWYNGkSAgMDjdafPn0aTU1NRutHjBiBwYMH4/jx4w/sr6GhAXV1dUYfIiIiejyZ9fbrHTt24LvvvsPJkyfv2VZeXo5evXpBrVYbrXdwcEB5efkD+1y5ciXeeecdU5dKRERE3ZDZzshcuXIF//u//4tt27aZdCZ6XFwcamtrxc+VK1dM1jcRERF1L2YLMqdPn0ZlZSV0Oh3kcjnkcjmOHDmCjz76CHK5HA4ODmhsbERNTY3RfhUVFdBoNA/s19LSEiqVyuhDREREjyezXVoKCAjA999/b7QuOjoaI0aMwOLFi+Ho6AgLCwtkZWUhMjISAFBYWIiSkhL4+PiYo2QiIiLqZswWZKytreHh4WG0rk+fPrCzsxPXz5kzB7Gxsejbty9UKhVee+01+Pj44OmnnzZHyURERNTNdOt3LSUmJqJHjx6IjIxEQ0MDgoODH/r+CSIiIvp56VZBJjs722jZysoK69evx/r1681TEBERdUsl747s0vEGL/v+0Y3+S1VVFZYtW4a9e/eioqICtra28PT0xLJly+Dre+d1MDKZDHv27EF4eHgnVNx2D6pn/vz52LhxI3bs2IGpU6eap7iH6FZBhoiI6HEQGRmJxsZGpKamYujQoaioqEBWVhaqq6vNXVqb3Lx5Ezt27MCiRYuwefPmbhlkzP5APCIiosdJTU0Njh49ilWrVmH8+PEYMmQIxo4di7i4OEyePBkAoNVqAQARERGQyWTiclFREcLCwuDg4AClUgkvLy9kZmYa9V9WVoZJkyZBoVDAyckJ27dvh1arxbp164xqiImJgb29PVQqFfz9/ZGfn9/mY7n7Fu233noLX3/9dbd8pAmDDBERkQkplUoolUqkp6ejoaHhvm3uPgg2OTkZZWVl4rLBYMDEiRORlZWFvLw8hISEIDQ0FCUlJeK+M2fORGlpKbKzs7Fr1y5s2LABlZWVRv1PnToVlZWV2L9/P06fPg2dToeAgABcu3atTceyadMmzJgxAzY2NpgwYQJSUlLatH9XYJAhIiIyIblcjpSUFKSmpkKtVsPX1xfx8fE4e/as2Mbe3h4AoFarodFoxGVPT0/Mnz8fHh4ecHFxwYoVKzBs2DBkZGQAAM6fP4/MzEwkJSXB29sbOp0OGzduxK1bt8S+v/nmG+Tm5iItLQ1jxoyBi4sLEhISoFarsXPnzlYfx8WLF3HixAm88MILAIAZM2YgOTkZgiB0+GdkSgwyREREJhYZGYnS0lJkZGQgJCQE2dnZ0Ol0jzyjYTAYsHDhQri5uUGtVkOpVKKgoEA8I1NYWAi5XA6dTifu4+zsbPTS5fz8fBgMBtjZ2Ylnh5RKJfR6PYqKilp9DJs3b0ZwcDD69esHAJg4cSJqa2tx+PDhNvwkOh8n+xIREXUCKysrBAUFISgoCEuXLkVMTAyWL1+OqKioB+6zcOFCHDp0CAkJCXB2doZCocCUKVPQ2NjY6nENBgMGDBhwz53AAO55f+GDNDc3IzU1FeXl5ZDL5UbrN2/ejICAgFbX09kYZIiIiLqAu7s70tPTxWULCws0NzcbtcnJyUFUVBQiIiIA3AklxcXF4nZXV1fcvn0beXl5GD16NADg0qVLuH79uthGp9OJAeTuJOK22rdvH27cuIG8vDz07NlTXP/DDz8gOjoaNTU1rQ5FnY2XloiIiEyouroa/v7+2Lp1K86ePQu9Xo+0tDSsXr0aYWFhYjutVousrCyUl5eLQcTFxQW7d+/GmTNnkJ+fj2nTpqGlpUXcZ8SIEQgMDMS8efOQm5uLvLw8zJs3DwqFAjKZDAAQGBgIHx8fhIeH4+DBgyguLsaxY8ewZMkSnDp1qlXHsGnTJkyaNAmenp7w8PAQP88//zzUajW2bdtmwp9YxzDIEBERmZBSqYS3tzcSExPh5+cHDw8PLF26FHPnzsUnn3witluzZg0OHToER0dH/OIXvwAArF27Fra2thg3bhxCQ0MRHBxsNB8GALZs2QIHBwf4+fkhIiICc+fOhbW1NaysrADcebDdvn374Ofnh+joaAwfPhwvvvgi/vWvf8HBweGR9VdUVGDv3r3iew7/W48ePRAREYFNmzZ15EdkUjKhu00/NrG6ujrY2NigtraWb8J+BN+PfTt9jJzXcjp9DCJ6fNTX10Ov18PJyUn8oiZjV69ehaOjIzIzM7vV3JXWeNjvt7Xf35wjQ0REJCGHDx+GwWDAyJEjUVZWhkWLFkGr1cLPz8/cpZkFgwwREZGENDU1IT4+HpcvX4a1tTXGjRuHbdu2wcLCwtylmQWDDBERkYQEBwcjODjY3GV0G5zsS0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLF58gQEZHkdMUrVf5bW1+vUlVVhWXLlmHv3r2oqKiAra0tPD09sWzZMvj63qldJpNhz549CA8P74SK2+6n9Tyovvnz52Pjxo3YsWMHpk6d2vWF/gSDDBERkYlFRkaisbERqampGDp0KCoqKpCVlYXq6mpzl9YhN2/exI4dO7Bo0SJs3ry5WwQZXloiIiIyoZqaGhw9ehSrVq3C+PHjMWTIEIwdOxZxcXGYPHkyAECr1QIAIiIiIJPJxOWioiKEhYXBwcEBSqUSXl5eyMzMNOq/rKwMkyZNgkKhgJOTE7Zv3w6tVot169YZ1RATEwN7e3uoVCr4+/sjPz+/w8eWlpYGd3d3vPXWW/j6669x5cqVDvfZUQwyREREJqRUKqFUKpGeno6Ghob7tjl58iQAIDk5GWVlZeKywWDAxIkTkZWVhby8PISEhCA0NBQlJSXivjNnzkRpaSmys7Oxa9cubNiwAZWVlUb9T506FZWVldi/fz9Onz4NnU6HgIAAXLt2rUPHtmnTJsyYMQM2NjaYMGECUlJSOtSfKTDIEBERmZBcLkdKSgpSU1OhVqvh6+uL+Ph4nD17Vmxjb28PAFCr1dBoNOKyp6cn5s+fDw8PD7i4uGDFihUYNmwYMjIyAADnz59HZmYmkpKS4O3tDZ1Oh40bN+LWrVti39988w1yc3ORlpaGMWPGwMXFBQkJCVCr1di5c2e7j+vixYs4ceIEXnjhBQDAjBkzkJycDEEQ2t2nKTDIEBERmVhkZCRKS0uRkZGBkJAQZGdnQ6fTPfIMhsFgwMKFC+Hm5ga1Wg2lUomCggLxjExhYSHkcjl0Op24j7OzM2xtbcXl/Px8GAwG2NnZiWeHlEol9Ho9ioqK2n1MmzdvRnBwMPr16wcAmDhxImpra3H48OF292kKnOxLRETUCaysrBAUFISgoCAsXboUMTExWL58OaKioh64z8KFC3Ho0CEkJCTA2dkZCoUCU6ZMQWNjY6vHNRgMGDBgALKzs+/Zplar234gAJqbm5Gamory8nLI5XKj9Zs3b0ZAQEC7+jUFBhkiIqIu4O7ujvT0dHHZwsICzc3NRm1ycnIQFRWFiIgIAHdCSXFxsbjd1dUVt2/fRl5eHkaPHg0AuHTpEq5fvy620el0YuC4O4m4o/bt24cbN24gLy8PPXv2FNf/8MMPiI6ORk1NTbtDUkfx0hIREZEJVVdXw9/fH1u3bsXZs2eh1+uRlpaG1atXIywsTGyn1WqRlZWF8vJyMYi4uLhg9+7dOHPmDPLz8zFt2jS0tLSI+4wYMQKBgYGYN28ecnNzkZeXh3nz5kGhUEAmkwEAAgMD4ePjg/DwcBw8eBDFxcU4duwYlixZglOnTrXrmDZt2oRJkybB09MTHh4e4uf555+HWq3Gtm3bOvAT6xiekSEiIslp6wPqupJSqYS3tzcSExNRVFSEpqYmODo6Yu7cuYiPjxfbrVmzBrGxsUhKSsKgQYNQXFyMtWvXYvbs2Rg3bhz69euHxYsXo66uzqj/LVu2YM6cOfDz84NGo8HKlStx7tw5WFlZAbjzILt9+/ZhyZIliI6ORlVVFTQaDfz8/ODg4NDm46moqMDevXuxffv2e7b16NEDERER2LRpExYsWNDmvk1BJph7unEnq6urg42NDWpra6FSqcxdTrfWFU/K7M7/8yGi7qe+vh56vR5OTk7iFzUZu3r1KhwdHZGZmWnWuSrt8bDfb2u/v3lGhoiISEIOHz4Mg8GAkSNHoqysDIsWLYJWq4Wfn5+5SzMLBhkiIiIJaWpqQnx8PC5fvgxra2uMGzcO27Ztg4WFhblLMwsGGSIiIgkJDg5GcHCwucvoNnjXEhEREUkWgwwRERFJFoMMERERSRaDDBEREUkWgwwRERFJFoMMERERSRaDDBEREUkWnyNDRESSc8TvmS4d75mvj7SpfVVVFZYtW4a9e/eioqICtra28PT0xLJly+Dre+d1MDKZDHv27EF4eHgnVNx2P63nfst39ezZEwMHDsSUKVOwcuVKWFpamqHiOxhkiIiITCwyMhKNjY1ITU3F0KFDUVFRgaysLFRXV5u7tA5JTk5GSEgImpqakJ+fj+joaPTp0wcrVqwwW028tERERGRCNTU1OHr0KFatWoXx48djyJAhGDt2LOLi4jB58mQAgFarBQBERERAJpOJy0VFRQgLC4ODgwOUSiW8vLyQmZlp1H9ZWRkmTZoEhUIBJycnbN++HVqtFuvWrTOqISYmBvb29lCpVPD390d+fn6Hj02tVkOj0cDR0RHPPfccwsLC8N1333W4345gkCEiIjIhpVIJpVKJ9PR0NDQ03LfNyZMnAdw5w1FWViYuGwwGTJw4EVlZWcjLy0NISAhCQ0NRUlIi7jtz5kyUlpYiOzsbu3btwoYNG1BZWWnU/9SpU1FZWYn9+/fj9OnT0Ol0CAgIwLVr10x2nBcuXMDhw4fh7e1tsj7bg0GGiIjIhORyOVJSUpCamgq1Wg1fX1/Ex8fj7NmzYht7e3sA/3eG4+6yp6cn5s+fDw8PD7i4uGDFihUYNmwYMjIyAADnz59HZmYmkpKS4O3tDZ1Oh40bN+LWrVti39988w1yc3ORlpaGMWPGwMXFBQkJCVCr1di5c2eHju2ll16CUqmElZUVXF1d8eSTTyIuLq5DfXYUgwwREZGJRUZGorS0FBkZGQgJCUF2djZ0Oh1SUlIeup/BYMDChQvh5uYGtVoNpVKJgoIC8YxMYWEh5HI5dDqduI+zszNsbW3F5fz8fBgMBtjZ2Ylnh5RKJfR6PYqKijp0XImJiThz5gzy8/Px97//HRcuXMBvfvObDvXZUZzsS0RE1AmsrKwQFBSEoKAgLF26FDExMVi+fDmioqIeuM/ChQtx6NAhJCQkwNnZGQqFAlOmTEFjY2OrxzUYDBgwYACys7Pv2aZWq9t+IP9Fo9HA2dkZAODq6oobN27gpZdewnvvvSeu72oMMkRERF3A3d0d6enp4rKFhQWam5uN2uTk5CAqKgoREREA7oSS4uJicburqytu376NvLw8jB49GgBw6dIlXL9+XWyj0+lQXl4OuVwuTiLuLD179gQAo0tbXY2XloiIiEyouroa/v7+2Lp1K86ePQu9Xo+0tDSsXr0aYWFhYjutVousrCyUl5eLQcTFxQW7d+8WL99MmzYNLS0t4j4jRoxAYGAg5s2bh9zcXOTl5WHevHlQKBTic14CAwPh4+OD8PBwHDx4EMXFxTh27BiWLFmCU6dOdejYampqUF5ejtLSUhw5cgTvvvsuhg8fDjc3tw712xE8I0NERJLT1gfUdSWlUglvb28kJiaiqKgITU1NcHR0xNy5cxEfHy+2W7NmDWJjY5GUlIRBgwahuLgYa9euxezZszFu3Dj069cPixcvRl1dnVH/W7ZswZw5c+Dn5weNRoOVK1fi3LlzsLKyAnDnwXX79u3DkiVLEB0djaqqKmg0Gvj5+cHBwaFDxxYdHS2OcbfPDz74AHK5+eKETBAEwWyjd4G6ujrY2NigtrYWKpXK3OV0a74f+3b6GDmv5XT6GET0+Kivr4der4eTk5P4RU3Grl69CkdHR2RmZiIgIMDc5bTJw36/rf3+5hkZIiIiCTl8+DAMBgNGjhyJsrIyLFq0CFqtFn5+fuYuzSwYZIiIiCSkqakJ8fHxuHz5MqytrTFu3Dhs27YNFhYW5i7NLBhkiIiIJCQ4OBjBwcHmLqPb4F1LREREJFkMMkRERCRZDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQkWQwyREREJFl8jgwREUnOJ29+2aXjvbomtE3tq6qqsGzZMuzduxcVFRWwtbWFp6cnli1bBl/fO6+Dkclk2LNnD8LDwzuh4rb7aT2PWu4uGGSIiIhMLDIyEo2NjUhNTcXQoUNRUVGBrKwsVFdXm7u0xw4vLREREZlQTU0Njh49ilWrVmH8+PEYMmQIxo4di7i4OEyePBkAoNVqAQARERGQyWTiclFREcLCwuDg4AClUgkvLy9kZmYa9V9WVoZJkyZBoVDAyckJ27dvh1arxbp164xqiImJgb29PVQqFfz9/ZGfn98Vh9/lGGSIiIhMSKlUQqlUIj09HQ0NDfdtc/LkSQBAcnIyysrKxGWDwYCJEyciKysLeXl5CAkJQWhoKEpKSsR9Z86cidLSUmRnZ2PXrl3YsGEDKisrjfqfOnUqKisrsX//fpw+fRo6nQ4BAQG4du1aJx21+TDIEBERmZBcLkdKSgpSU1OhVqvh6+uL+Ph4nD17Vmxjb28PAFCr1dBoNOKyp6cn5s+fDw8PD7i4uGDFihUYNmwYMjIyAADnz59HZmYmkpKS4O3tDZ1Oh40bN+LWrVti39988w1yc3ORlpaGMWPGwMXFBQkJCVCr1di5c2cX/iS6BoMMERGRiUVGRqK0tBQZGRkICQlBdnY2dDodUlJSHrqfwWDAwoUL4ebmBrVaDaVSiYKCAvGMTGFhIeRyOXQ6nbiPs7MzbG1txeX8/HwYDAbY2dmJZ4eUSiX0ej2Kioo65XjNiZN9iYiIOoGVlRWCgoIQFBSEpUuXIiYmBsuXL0dUVNQD91m4cCEOHTqEhIQEODs7Q6FQYMqUKWhsbGz1uAaDAQMGDEB2dvY929RqddsPpJsz6xmZzz77DKNGjYJKpYJKpYKPjw/2798vbq+vr8eCBQvEVBkZGYmKigozVkxERNQ+7u7u+PHHH8VlCwsLNDc3G7XJyclBVFQUIiIiMHLkSGg0GhQXF4vbXV1dcfv2beTl5YnrLl26hOvXr4vLOp0O5eXlkMvlcHZ2Nvr069ev8w7QTMwaZJ544gl8+OGHOH36NE6dOgV/f3+EhYXh3LlzAIA33ngDX375JdLS0nDkyBGUlpbi17/+tTlLJiIieqjq6mr4+/tj69atOHv2LPR6PdLS0rB69WqEhYWJ7bRaLbKyslBeXi4GERcXF+zevRtnzpxBfn4+pk2bhpaWFnGfESNGIDAwEPPmzUNubi7y8vIwb948KBQKyGQyAEBgYCB8fHwQHh6OgwcPori4GMeOHcOSJUtw6tSprv1hdAGzXloKDTV+wND777+Pzz77DCdOnMATTzyBTZs2Yfv27fD39wdwZ3a3m5sbTpw4gaefftocJRMRUTfQ1gfUdSWlUglvb28kJiaiqKgITU1NcHR0xNy5cxEfHy+2W7NmDWJjY5GUlIRBgwahuLgYa9euxezZszFu3Dj069cPixcvRl1dnVH/W7ZswZw5c+Dn5weNRoOVK1fi3LlzsLKyAnDnwXX79u3DkiVLEB0djaqqKmg0Gvj5+cHBwaFLfxZdQSYIgmDuIgCgubkZaWlpmDVrFvLy8lBeXo6AgABcv37d6JrekCFD8Prrr+ONN964bz8NDQ1Gt7vV1dXB0dERtbW1UKlUnX0Ykub7sW+nj5HzWk6nj0FEj4/6+nro9Xo4OTmJX9Rk7OrVq3B0dERmZiYCAgLMXU6bPOz3W1dXBxsbm0d+f5t9su/3338PHx8f1NfXQ6lUYs+ePXB3d8eZM2fQq1eveyYmOTg4oLy8/IH9rVy5Eu+8804nV01ERGQehw8fhsFgwMiRI1FWVoZFixZBq9XCz8/P3KWZhdlvv3Z1dcWZM2fw7bff4uWXX8asWbPwz3/+s939xcXFoba2VvxcuXLFhNUSERGZV1NTE+Lj4/Hkk08iIiIC9vb2yM7OhoWFhblLMwuzn5Hp1asXnJ2dAQCjR4/GyZMn8cc//hEvvPACGhsbUVNTY3RWpqKiAhqN5oH9WVpawtLSsrPLJiIiMovg4GAEBwebu4xuw+xnZH6qpaUFDQ0NGD16NCwsLJCVlSVuKywsRElJCXx8fMxYIREREXUXZj0jExcXhwkTJmDw4MG4ceMGtm/fjuzsbHz11VewsbHBnDlzEBsbi759+0KlUuG1116Dj48P71giIiIiAGYOMpWVlZg5cybKyspgY2ODUaNG4auvvkJQUBAAIDExET169EBkZCQaGhoQHByMTz/91JwlExERUTdi1iCzadOmh263srLC+vXrsX79+i6qiIiIiKSk282RISIiImotBhkiIiKSLAYZIiIikiyzP0eGiIiord6fMaVLx1uydWeb2ldVVWHZsmXYu3cvKioqYGtrC09PTyxbtgy+vndeByOTybBnzx6Eh4d3QsVt99N6HrV81/z587Fx40bs2LEDU6dO7dqiwSBDRERkcpGRkWhsbERqaiqGDh2KiooKZGVlobq62tylmdTNmzexY8cOLFq0CJs3bzZLkOGlJSIiIhOqqanB0aNHsWrVKowfPx5DhgzB2LFjERcXh8mTJwMAtFotACAiIgIymUxcLioqQlhYGBwcHKBUKuHl5YXMzEyj/svKyjBp0iQoFAo4OTlh+/bt0Gq1WLdunVENMTExsLe3h0qlgr+/P/Lz801+rGlpaXB3d8dbb72Fr7/+2iyvBWKQISIiMiGlUgmlUon09HQ0NDTct83JkycBAMnJySgrKxOXDQYDJk6ciKysLOTl5SEkJAShoaEoKSkR9505cyZKS0uRnZ2NXbt2YcOGDaisrDTqf+rUqaisrMT+/ftx+vRp6HQ6BAQE4Nq1ayY91k2bNmHGjBmwsbHBhAkTkJKSYtL+W4NBhoiIyITkcjlSUlKQmpoKtVoNX19fxMfH4+zZs2Ibe3t7AIBarYZGoxGXPT09MX/+fHh4eMDFxQUrVqzAsGHDkJGRAQA4f/48MjMzkZSUBG9vb+h0OmzcuBG3bt0S+/7mm2+Qm5uLtLQ0jBkzBi4uLkhISIBarcbOnW2b6/MwFy9exIkTJ/DCCy8AAGbMmIHk5GQIgmCyMVqDQYaIiMjEIiMjUVpaioyMDISEhCA7Oxs6ne6RZywMBgMWLlwINzc3qNVqKJVKFBQUiGdkCgsLIZfLodPpxH2cnZ1ha2srLufn58NgMMDOzk48O6RUKqHX61FUVGSyY9y8eTOCg4PRr18/AMDEiRNRW1uLw4cPm2yM1uBkXyIiok5gZWWFoKAgBAUFYenSpYiJicHy5csRFRX1wH0WLlyIQ4cOISEhAc7OzlAoFJgyZQoaGxtbPa7BYMCAAQOQnZ19zza1Wt32A7mP5uZmpKamory8HHK53Gj95s2bERAQYJJxWoNBhoiIqAu4u7sjPT1dXLawsEBzc7NRm5ycHERFRSEiIgLAnVBSXFwsbnd1dcXt27eRl5eH0aNHAwAuXbqE69evi210Op0YMO5OIja1ffv24caNG8jLy0PPnj3F9T/88AOio6NRU1NjstD0KLy0REREZELV1dXw9/fH1q1bcfbsWej1eqSlpWH16tUICwsT22m1WmRlZaG8vFwMIi4uLti9ezfOnDmD/Px8TJs2DS0tLeI+I0aMQGBgIObNm4fc3Fzk5eVh3rx5UCgUkMlkAIDAwED4+PggPDwcBw8eRHFxMY4dO4YlS5bg1KlTJjnGTZs2YdKkSfD09ISHh4f4ef7556FWq7Ft2zaTjNMaPCNDRESS09YH1HUlpVIJb29vJCYmoqioCE1NTXB0dMTcuXMRHx8vtluzZg1iY2ORlJSEQYMGobi4GGvXrsXs2bMxbtw49OvXD4sXL0ZdXZ1R/1u2bMGcOXPg5+cHjUaDlStX4ty5c7CysgJw58F1+/btw5IlSxAdHY2qqipoNBr4+fnBwcGhw8dXUVGBvXv3Yvv27fds69GjByIiIrBp0yYsWLCgw2O1hkzo6unFXayurg42Njaora2FSqUydzndmu/Hvp0+Rs5rOZ0+BhE9Purr66HX6+Hk5CR+UZOxq1evwtHREZmZmV06N8UUHvb7be33N8/IEBERScjhw4dhMBgwcuRIlJWVYdGiRdBqtfDz8zN3aWbBIENERCQhTU1NiI+Px+XLl2FtbY1x48Zh27ZtsLCwMHdpZsEgQ0REJCHBwcEIDg42dxndBu9aIiIiIslikCEiIiLJYpAhIiIiyWKQISIiIslikCEiIiLJYpAhIiIiyWKQISIiIsnic2SIiEhyCt4/3KXjuS3xb1P7qqoqLFu2DHv37kVFRQVsbW3h6emJZcuWwdf3zutgZDIZ9uzZg/Dw8E6ouO1+Wk9rlu/q2bMnBg4ciClTpmDlypWwtLTssroZZIiIiEwsMjISjY2NSE1NxdChQ1FRUYGsrCxUV1ebuzSTSk5ORkhICJqampCfn4/o6Gj06dMHK1as6LIaeGmJiIjIhGpqanD06FGsWrUK48ePx5AhQzB27FjExcVh8uTJAACtVgsAiIiIgEwmE5eLiooQFhYGBwcHKJVKeHl5ITMz06j/srIyTJo0CQqFAk5OTti+fTu0Wi3WrVtnVENMTAzs7e2hUqng7++P/Px8kx+rWq2GRqOBo6MjnnvuOYSFheG7774z+TgPwyBDRERkQkqlEkqlEunp6WhoaLhvm5MnTwK4c0ajrKxMXDYYDJg4cSKysrKQl5eHkJAQhIaGoqSkRNx35syZKC0tRXZ2Nnbt2oUNGzagsrLSqP+pU6eisrIS+/fvx+nTp6HT6RAQEIBr16510lEDFy5cwOHDh+Ht7d1pY9wPgwwREZEJyeVypKSkIDU1FWq1Gr6+voiPj8fZs2fFNvb29gD+74zG3WVPT0/Mnz8fHh4ecHFxwYoVKzBs2DBkZGQAAM6fP4/MzEwkJSXB29sbOp0OGzduxK1bt8S+v/nmG+Tm5iItLQ1jxoyBi4sLEhISoFarsXPnTpMe60svvQSlUgkrKyu4urriySefRFxcnEnHeBQGGSIiIhOLjIxEaWkpMjIyEBISguzsbOh0OqSkpDx0P4PBgIULF8LNzQ1qtRpKpRIFBQXiGZnCwkLI5XLodDpxH2dnZ9ja2orL+fn5MBgMsLOzE88OKZVK6PV6FBUVmfQ4ExMTcebMGeTn5+Pvf/87Lly4gN/85jcmHeNRONmXiIioE1hZWSEoKAhBQUFYunQpYmJisHz5ckRFRT1wn4ULF+LQoUNISEiAs7MzFAoFpkyZgsbGxlaPazAYMGDAAGRnZ9+zTa1Wt/1AHkKj0cDZ2RkA4Orqihs3buCll17Ce++9J67vbO06I+Pv74+ampp71tfV1cHfv223qBEREf0cuLu748cffxSXLSws0NzcbNQmJycHUVFRiIiIwMiRI6HRaFBcXCxud3V1xe3bt5GXlyeuu3TpEq5fvy4u63Q6lJeXQy6Xw9nZ2ejTr1+/zjtA3LkNG4DRpa7O1q4gk52dfd90WF9fj6NHj3a4KCIiIqmqrq6Gv78/tm7dirNnz0Kv1yMtLQ2rV69GWFiY2E6r1SIrKwvl5eViEHFxccHu3bvFyzXTpk1DS0uLuM+IESMQGBiIefPmITc3F3l5eZg3bx4UCoX4XJfAwED4+PggPDwcBw8eRHFxMY4dO4YlS5bg1KlTJj3WmpoalJeXo7S0FEeOHMG7776L4cOHw83NzaTjPEybLi3990Slf/7znygvLxeXm5ubceDAAQwaNMh01REREd1HWx9Q15WUSiW8vb2RmJiIoqIiNDU1wdHREXPnzkV8fLzYbs2aNYiNjUVSUhIGDRqE4uJirF27FrNnz8a4cePQr18/LF68GHV1dUb9b9myBXPmzIGfnx80Gg1WrlyJc+fOwcrKCsCdB9Xt27cPS5YsQXR0NKqqqqDRaODn5wcHBweTHmt0dLQ45t0xPvjgA8jlXTdzRSYIgtDaxj169BAT3/12UygU+PjjjzF79mzTVdhBdXV1sLGxQW1tLVQqlbnL6dZ8P/bt9DFyXsvp9DGI6PFRX18PvV4PJycn8YuajF29ehWOjo7IzMxEQECAuctpk4f9flv7/d2myKTX6yEIAoYOHYrc3FzxdjEA6NWrF/r37y9eHyMiIiLTO3z4MAwGA0aOHImysjIsWrQIWq0Wfn5+5i7NLNoUZIYMGQIARtfriIiIqOs0NTUhPj4ely9fhrW1NcaNG4dt27bBwsLC3KWZRbsvYl28eBH/+Mc/UFlZeU+wWbZsWYcLIyIionsFBwcjODjY3GV0G+0KMklJSXj55ZfRr18/aDQaozdgymQyBhkiIiLqEu0KMu+99x7ef/99LF682NT1EBEREbVau54jc/36dUydOtXUtRARERG1SbuCzNSpU3Hw4EFT10JERETUJu26tOTs7IylS5fixIkTGDly5D0zpX/729+apDgiIiKih2lXkNmwYQOUSiWOHDmCI0eOGG2TyWQMMkRERNQl2hVk9Hq9qesgIiJqtbfffvuxHo9ar11zZIiIiOjBqqqq8PLLL2Pw4MGwtLSERqNBcHAwcnL+7zUtMpkM6enp5ivyJ35az6OWAeAf//gHJk6cCDs7O/Tu3Rvu7u5488038e9//7trikY7z8g86l1KmzdvblcxREREj4PIyEg0NjYiNTUVQ4cORUVFBbKyslBdXW3u0kzmiy++wCuvvIJZs2Zh165d0Gq1KCkpwZYtW7BmzRqsXbu2S+po9+3X//2prKzE4cOHsXv3btTU1Ji4RCIiIumoqanB0aNHsWrVKowfPx5DhgzB2LFjERcXh8mTJwMAtFotACAiIgIymUxcLioqQlhYGBwcHKBUKuHl5YXMzEyj/svKyjBp0iQoFAo4OTlh+/bt0Gq1WLdunVENMTExsLe3h0qlgr+/P/Lz8012jFevXsVvf/tb/Pa3v8XmzZvx7LPPiu972rhxY5c+GLddZ2T27Nlzz7qWlha8/PLLGDZsWIeLIiIikiqlUgmlUon09HQ8/fTTsLS0vKfNyZMn0b9/fyQnJyMkJER84bLBYMDEiRPx/vvvw9LSElu2bEFoaCgKCwsxePBgAMDMmTPxn//8B9nZ2bCwsEBsbCwqKyuN+p86dSoUCgX2798PGxsbfPHFFwgICMCFCxfQt2/fDh9jWloaGhsbsWjRovtuV6vVHR6jtUw2R6ZHjx6IjY1FYmKiqbokIiKSHLlcjpSUFKSmpkKtVsPX1xfx8fE4e/as2Mbe3h7AnS98jUYjLnt6emL+/Pnw8PCAi4sLVqxYgWHDhiEjIwMAcP78eWRmZiIpKQne3t7Q6XTYuHEjbt26Jfb9zTffIDc3F2lpaRgzZgxcXFyQkJAAtVqNnTt3muQYL168CJVKhQEDBpikv44w6WTfoqIi3L5925RdEhERSU5kZCRKS0uRkZGBkJAQZGdnQ6fTISUl5aH7GQwGLFy4EG5ublCr1VAqlSgoKEBJSQkAoLCwEHK5HDqdTtzH2dkZtra24nJ+fj4MBgPs7OzEs0NKpRJ6vR5FRUUmOT5BEIzes2hO7bq0FBsba7QsCALKysqwd+9ezJo1yySFERERSZmVlRWCgoIQFBSEpUuXIiYmBsuXL0dUVNQD91m4cCEOHTqEhIQEODs7Q6FQYMqUKWhsbGz1uAaDAQMGDEB2dvY920x1yWf48OGora1FWVmZ2c/KtCvI5OXlGS336NED9vb2WLNmzSPvaCIiIvo5cnd3N7p92cLCAs3NzUZtcnJyEBUVhYiICAB3QklxcbG43dXVFbdv30ZeXh5Gjx4NALh06RKuX78uttHpdCgvL4dcLhcnEZvalClT8NZbb2H16tX3nVJSU1PTZfNk2hVk/vGPf5i6DiIiosdCdXU1pk6ditmzZ2PUqFGwtrbGqVOnsHr1aoSFhYnttFotsrKy4OvrC0tLS9ja2sLFxQW7d+9GaGgoZDIZli5dipaWFnGfESNGIDAwEPPmzcNnn30GCwsLvPnmm1AoFOKlnsDAQPj4+CA8PByrV6/G8OHDUVpair179yIiIgJjxozp8DE6OjoiMTERr776Kurq6jBz5kxotVpcvXoVW7ZsgVKpxJo1azo8Tmu0K8jcVVVVhcLCQgB3UuLdyUpERESdqTs/aVepVMLb2xuJiYkoKipCU1MTHB0dMXfuXMTHx4vt1qxZg9jYWCQlJWHQoEEoLi7G2rVrMXv2bIwbNw79+vXD4sWLUVdXZ9T/li1bMGfOHPj5+UGj0WDlypU4d+4crKysANx5cN2+ffuwZMkSREdHo6qqChqNBn5+fnBwcDDZcb7yyisYPnw4EhISEBERgVu3bkGr1eK55567ZwpKZ5IJgiC0dacff/wRr732GrZs2SImxZ49e2LmzJn4+OOP0bt3b5MX2l51dXWwsbFBbW0tVCqVucvp1nw/9u30MXJey3l0IyKi/199fT30ej2cnJzEL2oydvXqVTg6OiIzMxMBAQHmLqdNHvb7be33d7vuWoqNjcWRI0fw5ZdfoqamBjU1Nfjb3/6GI0eO4M0332xPl0RERNQKhw8fRkZGBvR6PY4dO4YXX3xRfBjdz1G7Li3t2rULO3fuxLPPPiuumzhxIhQKBZ5//nl89tlnpqqPiIiI/ktTUxPi4+Nx+fJlWFtbY9y4cdi2bRssLCzMXZpZtCvI3Lx5877X2fr374+bN292uCgiIiK6v+DgYAQHB5u7jG6jXZeWfHx8sHz5ctTX14vrbt26hXfeeQc+Pj4mK46IiIjoYdp1RmbdunUICQnBE088AU9PTwB3niRoaWmJgwcPmrRAIiKidtyXQhJgit9ru4LMyJEjcfHiRWzbtg3nz58HALz00kuYPn06FApFh4siIiICIM77uHnzJr9fHkN3p6N0ZH5Pu4LMypUr4eDggLlz5xqt37x5M6qqqrB48eJ2F0RERHRXz549oVarxbc79+7du9u844faTxAE3Lx5E5WVlVCr1eLbv9ujXUHmiy++wPbt2+9Z/+STT+LFF19kkCEiIpPRaDQAIIYZenzcfft3R7QryJSXl9/3JVH29vYoKyvrUEFERET/TSaTYcCAAejfvz+amprMXQ6ZiIWFRYfOxNzVriDj6OiInJwcODk5Ga3PycnBwIEDO1wUERHRT/Xs2dMkX3z0eGlXkJk7dy5ef/11NDU1wd/fHwCQlZWFRYsW8cm+RERE1GXaFWR+97vfobq6Gq+88goaGxsBAFZWVli8eDHi4uJMWiARERHRg7QryMhkMqxatQpLly5FQUEBFAoFXFxcYGlpaer6iIiIiB6oXU/2vUupVMLLywseHh7tCjErV66El5cXrK2t0b9/f4SHh6OwsNCoTX19PRYsWAA7OzsolUpERkaioqKiI2UTERHRY6JDQaajjhw5ggULFuDEiRM4dOgQmpqa8Ktf/Qo//vij2OaNN97Al19+ibS0NBw5cgSlpaX49a9/bcaqiYiIqLto16UlUzlw4IDRckpKCvr374/Tp0/Dz88PtbW12LRpE7Zv3y5OKk5OToabmxtOnDiBp59+2hxlExERUTdh1jMyP1VbWwsA6Nu3LwDg9OnTaGpqQmBgoNhmxIgRGDx4MI4fP37fPhoaGlBXV2f0ISIiosdTtwkyLS0teP311+Hr6wsPDw8Adx6816tXL6jVaqO2Dg4OKC8vv28/K1euhI2NjfhxdHTs7NKJiIjITLpNkFmwYAF++OEH7Nixo0P9xMXFoba2VvxcuXLFRBUSERFRd2PWOTJ3vfrqq/j73/+Or7/+Gk888YS4XqPRoLGxETU1NUZnZSoqKh74bgZLS0veBk5ERPQzYdYzMoIg4NVXX8WePXtw+PDhe155MHr0aFhYWCArK0tcV1hYiJKSEvj4+HR1uURERNTNmPWMzIIFC7B9+3b87W9/g7W1tTjvxcbGBgqFAjY2NpgzZw5iY2PRt29fqFQqvPbaa/Dx8eEdS0RERGTeIPPZZ58BAJ599lmj9cnJyYiKigIAJCYmokePHoiMjERDQwOCg4Px6aefdnGlRERE1B2ZNcgIgvDINlZWVli/fj3Wr1/fBRURERGRlHSbu5aIiIiI2opBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgkq1u8/ZpIat6fMaXTx1iydWenj0FEJHU8I0NERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREkiU3dwGPi9G/29LpY5z+w8xOH4OIiEhKeEaGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCRLbu4C6OfliN8znT7GM18f6fQxiIioe+AZGSIiIpIsBhkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLLMGma+//hqhoaEYOHAgZDIZ0tPTjbYLgoBly5ZhwIABUCgUCAwMxMWLF81TLBEREXU7Zg0yP/74Izw9PbF+/fr7bl+9ejU++ugjfP755/j222/Rp08fBAcHo76+vosrJSIiou7IrG+/njBhAiZMmHDfbYIgYN26dfj973+PsLAwAMCWLVvg4OCA9PR0vPjii11ZKhEREXVD3XaOjF6vR3l5OQIDA8V1NjY28Pb2xvHjxx+4X0NDA+rq6ow+RERE9HjqtkGmvLwcAODg4GC03sHBQdx2PytXroSNjY34cXR07NQ6iYiIyHy6bZBpr7i4ONTW1oqfK1eumLskIiIi6iTdNshoNBoAQEVFhdH6iooKcdv9WFpaQqVSGX2IiIjo8dRtg4yTkxM0Gg2ysrLEdXV1dfj222/h4+NjxsqIiIiouzDrXUsGgwGXLl0Sl/V6Pc6cOYO+ffti8ODBeP311/Hee+/BxcUFTk5OWLp0KQYOHIjw8HDzFU1ERETdhlmDzKlTpzB+/HhxOTY2FgAwa9YspKSkYNGiRfjxxx8xb9481NTU4Je//CUOHDgAKysrc5VMRERE3YhZg8yzzz4LQRAeuF0mk+Hdd9/Fu+++24VVERERkVR02zkyRERERI/CIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREksUgQ0RERJLFIENERESSxSBDREREkmXWJ/sSEdHj44jfM50+xjNfH+n0MUhaeEaGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgki0GGiIiIJItBhoiIiCSLQYaIiIgkS27uAoiIqPP5fuzb6WN8wK8UMgOekSEiIiLJYpAhIiIiyWKQISIiIslikCEiIiLJYpAhIiIiyWKQISIiIslikCEiIiLJYpAhIiIiyWKQISIiIslikCEiIiLJYpAhIiIiyWKQISIiIslikCEiIiLJYpAhIiIiyWKQISIiIsmSm7sAar2Sd0d27gC2qs7tn4hIAt6fMaXTx1iydWenj/FzwTMyREREJFkMMkRERCRZDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQkWQwyREREJFkMMkRERCRZDDJEREQkWXJzF0Bkap+8+aW5SzCJgvcPd2r/f2n6ulP7B4C3336708egn5fH5e+7s3XF3153+fvmGRkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLAYZIiIikiwGGSIiIpIsBhkiIiKSLD5HhoiIqIt19nOifk54RoaIiIgki0GGiIiIJItBhoiIiCRLEkFm/fr10Gq1sLKygre3N3Jzc81dEhEREXUD3T7I/OUvf0FsbCyWL1+O7777Dp6enggODkZlZaW5SyMiIiIz6/ZBZu3atZg7dy6io6Ph7u6Ozz//HL1798bmzZvNXRoRERGZWbe+/bqxsRGnT59GXFycuK5Hjx4IDAzE8ePH77tPQ0MDGhoaxOXa2loAQF1dXafW2txwq1P7B4AbFs2d2v/tW7c7tX8A+LHzh8CthpudPkZ9U1Onj2Go/7FT+2+43fDoRh3U2X931Hr8+249/n23Tmf/fd/tXxCEhzcUurF///vfAgDh2LFjRut/97vfCWPHjr3vPsuXLxcA8MMPP/zwww8/j8HnypUrD80K3fqMTHvExcUhNjZWXG5pacG1a9dgZ2cHmUxmxsqoK9TV1cHR0RFXrlyBSqUydzlEZEL8+/55EQQBN27cwMCBAx/arlsHmX79+qFnz56oqKgwWl9RUQGNRnPffSwtLWFpaWm0Tq1Wd1aJ1E2pVCr+j47oMcW/758PGxubR7bp1pN9e/XqhdGjRyMrK0tc19LSgqysLPj4+JixMiIiIuoOuvUZGQCIjY3FrFmzMGbMGIwdOxbr1q3Djz/+iOjoaHOXRkRERGbW7YPMCy+8gKqqKixbtgzl5eV46qmncODAATg4OJi7NOqGLC0tsXz58nsuLxKR9PHvm+5HJgiPuq+JiIiIqHvq1nNkiIiIiB6GQYaIiIgki0GGiIiIJItBhh47Wq0W69ate2gbmUyG9PT0LqmHiIg6D4MMScqVK1cwe/ZsDBw4EL169cKQIUPwv//7v6iurjZ3aUT0E4IgIDAwEMHBwfds+/TTT6FWq3H16lUzVEaPEwYZkozLly9jzJgxuHjxIv785z/j0qVL+Pzzz8UHJF67ds3cJRLRf5HJZEhOTsa3336LL774Qlyv1+uxaNEifPzxx3jiiSfMWCE9DhhkSDIWLFiAXr164eDBg3jmmWcwePBgTJgwAZmZmfj3v/+NJUuW3He/ixcvws/PD1ZWVnB3d8ehQ4e6uHKiny9HR0f88Y9/xMKFC6HX6yEIAubMmYNf/epXGDx4MMaOHQtLS0sMGDAAb731Fm7f/r9XaN/vMvFTTz2Ft99+W1yWyWTYuHEjIiIi0Lt3b7i4uCAjI8Non4yMDLi4uMDKygrjx49HamoqZDIZampqOvHIqaswyJAkXLt2DV999RVeeeUVKBQKo20ajQbTp0/HX/7yl3te997S0oJf//rX6NWrF7799lt8/vnnWLx4cVeWTvSzN2vWLAQEBGD27Nn45JNP8MMPPyAxMRETJ06El5cX8vPz8dlnn2HTpk1477332tz/O++8g+effx5nz57FxIkTMX36dPEMrV6vx5QpUxAeHo78/HzMnz//gf/oIWnq9k/2JQLunFURBAFubm733e7m5obr16+jqqrKaH1mZibOnz+Pr776SnyD6gcffIAJEyZ0es1E9H82bNiAJ598El9//TV27dqFDRs2wNHREZ988glkMhlGjBiB0tJSLF68GMuWLUOPHq3/d3ZUVBReeuklAHf+vj/66CPk5uYiJCQEX3zxBVxdXfGHP/wBAODq6ooffvgB77//fqccJ3U9npEhSWnrg6gLCgrg6Oho9Bp4vnCUqOv1798f8+fPh5ubG8LDw1FQUAAfHx/IZDKxja+vLwwGQ5snAI8aNUr87z59+kClUqGyshIAUFhYCC8vL6P2Y8eO7cCRUHfDIEOS4OzsDJlMhoKCgvtuLygogK2tLezt7bu4MiJqLblcDrm89RcCevTocc8/Xpqamu5pZ2FhYbQsk8nQ0tLSviJJchhkSBLs7OwQFBSETz/9FLdu3TLaVl5ejm3btuGFF14w+tcdcOeS05UrV1BWViauO3HiRJfUTEQP5ubmhuPHjxsFlZycHFhbW4t3Mtnb2xv97dbV1UGv17dpHFdXV5w6dcpo3cmTJztQOXU3DDIkGZ988gkaGhoQHByMr7/+GleuXMGBAwcQFBSEQYMG3fead2BgIIYPH45Zs2YhPz8fR48e5UQ/om7glVdewZUrV/Daa6/h/Pnz+Nvf/obly5cjNjZWnB/j7++PP/3pTzh69Ci+//57zJo1Cz179mzTOPPnz8f58+exePFiXLhwAX/961+RkpICAPf8w4ekiUGGJMPFxQWnTp3C0KFD8fzzz2PYsGGYN28exo8fj+PHj6Nv37737NOjRw/s2bMHt27dwtixYxETE8NJfkTdwKBBg7Bv3z7k5ubC09MT//M//4M5c+bg97//vdgmLi4OzzzzDJ577jlMmjQJ4eHhGDZsWJvGcXJyws6dO7F7926MGjUKn332mfiPGUtLS5MeE5mHTGjr7EkiIiIJe//99/H555/jypUr5i6FTIC3XxMR0WPt008/hZeXF+zs7JCTk4M//OEPePXVV81dFpkIgwwRET3WLl68iPfeew/Xrl3D4MGD8eabbyIuLs7cZZGJ8NISERERSRYn+xIREZFkMcgQERGRZDHIEBERkWQxyBAREZFkMcgQERGRZDHIEBERkWQxyBDRY+Htt9/GU089Ze4yiKiLMcgQEXWCpqYmc5dA9LPAIENEAIADBw7gl7/8JdRqNezs7PDcc8+hqKhI3H7s2DE89dRTsLKywpgxY5Ceng6ZTIYzZ86IbX744QdMmDABSqUSDg4O+M1vfoP//Oc/rRq/paUFq1evhrOzMywtLTF48GCjF3wuXrwYw4cPR+/evTF06FAsXbpUDAspKSl45513kJ+fD5lMBplMJr7huKamBjExMbC3t4dKpYK/vz/y8/ONxn7vvffQv39/WFtbIyYmBm+99ZbR2Z2Wlha8++67eOKJJ2BpaYmnnnoKBw4cELcXFxdDJpPhL3/5C5555hlYWVlhw4YNUKlU2Llzp9FY6enp6NOnD27cuNGqnwsRPYJARCQIws6dO4Vdu3YJFy9eFPLy8oTQ0FBh5MiRQnNzs1BbWyv07dtXmDFjhnDu3Dlh3759wvDhwwUAQl5eniAIgnD9+nXB3t5eiIuLEwoKCoTvvvtOCAoKEsaPH9+q8RctWiTY2toKKSkpwqVLl4SjR48KSUlJ4vYVK1YIOTk5gl6vFzIyMgQHBwdh1apVgiAIws2bN4U333xTePLJJ4WysjKhrKxMuHnzpiAIghAYGCiEhoYKJ0+eFC5cuCC8+eabgp2dnVBdXS0IgiBs3bpVsLKyEjZv3iwUFhYK77zzjqBSqQRPT09x7LVr1woqlUr485//LJw/f15YtGiRYGFhIVy4cEEQBEHQ6/UCAEGr1Qq7du0SLl++LJSWlgpz584VJk6caHSckydPFmbOnNmu3xER3YtBhojuq6qqSgAgfP/998Jnn30m2NnZCbdu3RK3JyUlGQWZFStWCL/61a+M+rhy5YoAQCgsLHzoWHV1dYKlpaVRcHmUP/zhD8Lo0aPF5eXLlxuFD0EQhKNHjwoqlUqor683Wj9s2DDhiy++EARBELy9vYUFCxYYbff19TXqa+DAgcL7779v1MbLy0t45ZVXBEH4vyCzbt06ozbffvut0LNnT6G0tFQQBEGoqKgQ5HK5kJ2d3erjJKKH46UlIgJw58V6L730EoYOHQqVSgWtVgsAKCkpQWFhIUaNGgUrKyux/dixY432z8/Pxz/+8Q8olUrxM2LECAAwukR1PwUFBWhoaEBAQMAD2/zlL3+Br68vNBoNlEolfv/736OkpOSh/ebn58NgMMDOzs6oLr1eL9ZUWFh4z7H893JdXR1KS0vh6+tr1MbX1xcFBQVG68aMGXNPP08++SRSU1MBAFu3bsWQIUPg5+f30LqJqPX49msiAgCEhoZiyJAhSEpKwsCBA9HS0gIPDw80Nja2an+DwYDQ0FCsWrXqnm0DBgx46L4KheKh248fP47p06fjnXfeQXBwMGxsbLBjxw6sWbPmkTUNGDAA2dnZ92xTq9UP3bc9+vTpc8+6mJgYrF+/Hm+99RaSk5MRHR0NmUxm8rGJfq54RoaIUF1djcLCQvz+979HQEAA3NzccP36dXG7q6srvv/+ezQ0NIjrTp48adSHTqfDuXPnoNVq4ezsbPS53xf8f3NxcYFCoUBWVtZ9tx87dgxDhgzBkiVLMGbMGLi4uOBf//qXUZtevXqhubn5nprKy8shl8vvqalfv37isf30WP57WaVSYeDAgcjJyTFqk5OTA3d394ceFwDMmDED//rXv/DRRx/hn//8J2bNmvXIfYioDcx9bYuIzK+5uVmws7MTZsyYIVy8eFHIysoSvLy8BADCnj17xMm+M2fOFP75z38KBw4cEEaMGCEAEM6cOSMIgiD8+9//Fuzt7YUpU6YIubm5wqVLl4QDBw4IUVFRwu3btx9Zw9tvvy3Y2toKqampwqVLl4Tjx48LGzduFARBEP72t78Jcrlc+POf/yxcunRJ+OMf/yj07dtXsLGxEffftm2b0KdPHyEvL0+oqqoS6uvrhZaWFuGXv/yl4OnpKXz11VeCXq8XcnJyhPj4eOHkyZOCINyZ7KtQKISUlBThwoULwooVKwSVSiU89dRTYt+JiYmCSqUSduzYIZw/f15YvHjxfSf73p0v9FPTpk0TevXqJYSEhLTn10NED8EgQ0SCIAjCoUOHBDc3N8HS0lIYNWqUkJ2dLQYZQRCEnJwcYdSoUUKvXr2E0aNHC9u3bxcACOfPnxf7uHDhghARESGo1WpBoVAII0aMEF5//XWhpaXlkeM3NzcL7733njBkyBDBwsJCGDx4sPDBBx+I23/3u98JdnZ2glKpFF544QUhMTHRKMjU19cLkZGRglqtFgAIycnJgiDcmUj82muvCQMHDhQsLCwER0dHYfr06UJJSYm477vvviv069dPUCqVwuzZs4Xf/va3wtNPP21U29tvvy0MGjRIsLCwEDw9PYX9+/eL2x8VZLKysgQAwl//+tdH/hyIqG1kgiAI5jsfRERStW3bNkRHR6O2tvaRc1ykJigoCBqNBn/6059M0t+f/vQnvPHGGygtLUWvXr1M0icR3cHJvkTUKlu2bMHQoUMxaNAg5OfnY/HixXj++eclH2Ju3ryJzz//HMHBwejZsyf+/Oc/IzMzE4cOHTJJ32VlZfjwww8xf/58hhiiTsDJvkTUKuXl5ZgxYwbc3NzwxhtvYOrUqdiwYUOr9i0pKTG6/fmnn0fdRt2ZZDIZ9u3bBz8/P4wePRpffvkldu3ahcDAwA73vXr1aowYMQIajQZxcXEmqJaIfoqXloio092+fRvFxcUP3K7VaiGX8wQxEbUdgwwRERFJFi8tERERkWQxyBAREZFkMcgQERGRZDHIEBERkWQxyBAREZFkMcgQERGRZDHIEBERkWT9f8qO5wuaBwRZAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 1. Import libraries here\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# 2. Create the age_category column in clinical_data\n",
    "clinical_data[\"age_category\"] = np.where(clinical_data.loc[:, \"Age\"] < 50, \"Young\", \"Old\")\n",
    "\n",
    "# 3. Replace NaN with No cancer\n",
    "clinical_data.loc[:, \"Stage\"] = np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "\n",
    "stage_categories = np.unique(clinical_data.loc[:,\"Stage\"])\n",
    "assert(np.all(stage_categories == ['No cancer', 'Stage IA', 'Stage IIA', 'Stage IIB',\n",
    "                                    'Stage III', 'Stage IIIA', 'Stage IIIB', 'Stage IIIC']))\n",
    "# 4. Loop through all the stage categories and count\n",
    "\n",
    "for age_cat in [\"Old\", \"Young\"]: # first get the old data, then the young data\n",
    "    print(age_cat)\n",
    "    age_mask = np.where(clinical_data.loc[:, \"age_category\"] == age_cat, True, False)\n",
    "    temp_dataFrame = pd.DataFrame(clinical_data[age_mask])\n",
    "    # fill this in with the cancer category:\n",
    "    for stage_category in stage_categories:\n",
    "        people = len(temp_dataFrame[temp_dataFrame[\"Stage\"] == stage_category])\n",
    "        print(f\"Category is {age_cat}, Stage is {stage_category}, Number is {people}\")\n",
    "\n",
    "#\n",
    "# 5. A sample barplot to compare your results\n",
    "clinical_data_copy = clinical_data\n",
    "clinical_data_copy[\"Stage\"] = np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "sns.countplot(\n",
    "    x = \"age_category\",\n",
    "    hue = \"Stage\",\n",
    "    hue_order = stage_categories,\n",
    "    data = clinical_data_copy\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "652509b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
